<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Methods for Access and Analysis of the Human Leukocyte Antigen (HLA) Gene Sequence Alignments from IMGTHLA • hlabud</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Methods for Access and Analysis of the Human Leukocyte Antigen (HLA) Gene Sequence Alignments from IMGTHLA">
<meta name="description" content="Fetch sequence alignment data from the IMGTHLA database by Robinson et al (2020) &lt;doi:10.1093/nar/gkz950&gt;, automatically convert the sequence alignments into convenient R matrices ready for downstream analysis. The vignette shows examples of using the one-hot encoding of the data with logistic regression and dimensionality reduction. Data is downloaded lazily, as-needed, and cached in a user-configurable folder.">
<meta property="og:description" content="Fetch sequence alignment data from the IMGTHLA database by Robinson et al (2020) &lt;doi:10.1093/nar/gkz950&gt;, automatically convert the sequence alignments into convenient R matrices ready for downstream analysis. The vignette shows examples of using the one-hot encoding of the data with logistic regression and dimensionality reduction. Data is downloaded lazily, as-needed, and cached in a user-configurable folder.">
<meta name="twitter:card" content="summary_large_image">
<meta name="twitter:creator" content="@slowkow">
<meta name="twitter:site" content="@slowkow">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">hlabud</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="articles/examples.html">hlabud usage examples</a></li>
    <li><a class="dropdown-item" href="articles/numbering.html">Numbering amino acid positions</a></li>
    <li><a class="dropdown-item" href="articles/visualize-hla-structure.html">Visualize HLA protein structures</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/slowkow/hlabud"><span class="fa fa-github"></span> GitHub</a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="hlabud-hla-analysis-in-r-">hlabud: HLA analysis in R <img width="25%" align="right" src="https://github.com/slowkow/hlabud/assets/209714/b39a3f04-c9a8-4867-a3e0-9434f0f9ef20"><a class="anchor" aria-label="anchor" href="#hlabud-hla-analysis-in-r-"></a>
</h1></div>
<p>hlabud provides methods to retrieve sequence alignment data from <a href="https://github.com/ANHIG/IMGTHLA" class="external-link">IMGTHLA</a> and convert the data into convenient R matrices ready for downstream analysis. See the <a href="https://slowkow.github.io/hlabud/articles/examples.html">usage examples</a> to learn how to use the data with logistic regression and dimensionality reduction.</p>
<p>For example, let’s consider a simple question about two HLA genotypes.</p>
<p>What amino acid positions are different between these two genotypes?</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">library</span>(hlabud)</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a>a <span class="ot">&lt;-</span> <span class="fu">hla_alignments</span>(<span class="st">"DRB1"</span>)</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="fu">dosage</span>(a<span class="sc">$</span>onehot, <span class="fu">c</span>(<span class="st">"DRB1*03:01:05"</span>, <span class="st">"DRB1*03:02:03"</span>))</span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a></span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a><span class="do">##               F26 Y26 D28 E28 F47 Y47 G86 V86</span></span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a><span class="do">## DRB1*03:01:05   0   1   1   0   1   0   0   1</span></span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a><span class="do">## DRB1*03:02:03   1   0   0   1   0   1   1   0</span></span></code></pre></div>
<p>From this output, we can conclude that four positions (26, 28, 47, 86) distinguish these two HLA-DRB1 alleles. We see that DRB1*03:01:05 has a Y at position 26 and DRB1*03:02:03 has a F.</p>
</div>
<div class="section level1">
<h1 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h1>
<p>The quickest way to get hlabud is to install from GitHub:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="co"># install.packages("devtools")</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">"slowkow/hlabud"</span>)</span></code></pre></div>
</div>
<div class="section level1">
<h1 id="examples">Examples<a class="anchor" aria-label="anchor" href="#examples"></a>
</h1>
<p>See the <a href="articles/examples.html">usage examples</a> to get some ideas for how to use hlabud in your analyses.</p>
<ul>
<li><p><a href="articles/examples.html#get-a-one-hot-encoded-matrix-for-all-hla-drb1-alleles">Get a one-hot encoded matrix for all HLA-DRB1 alleles</a></p></li>
<li><p><a href="articles/examples.html#convert-genotypes-to-a-dosage-matrix">Convert genotypes to a dosage matrix</a></p></li>
<li><p><a href="articles/examples.html#logistic-regression-association-for-amino-acid-positions">Logistic regression association for amino acid positions</a></p></li>
<li><p><a href="articles/examples.html#umap-embedding-of-hla-drb1-alleles">UMAP embedding of 3,671 HLA-DRB1 alleles</a></p></li>
<li><p><a href="articles/examples.html#get-hla-allele-frequencies-from-allele-frequency-net-database-afnd">Get HLA allele frequencies from Allele Frequency Net Database (AFND)</a></p></li>
<li><p><a href="articles/examples.html#compute-hla-divergence-with-the-grantham-distance-matrix">Compute HLA divergence with the Grantham distance matrix</a></p></li>
<li><p><a href="articles/examples.html#download-and-unpack-all-data-from-the-latest-imgthla-release">Download and unpack all data from the latest IMGTHLA release</a></p></li>
<li><p><a href="articles/visualize-hla-structure.html">Visualize the 3D molecular structure of HLA proteins and highlight specific amino acid residues</a></p></li>
</ul>
<p><a href="https://slowkow.github.io/hlabud/articles/examples.html#logistic-regression-association-for-amino-acid-positions"> <img width="49%" src="articles/examples_files/figure-html/glm-volcano-1.png"></a> <a href="https://slowkow.github.io/hlabud/articles/examples.html#umap-embedding-of-hla-drb1-alleles"> <img width="49%" src="articles/examples_files/figure-html/umap-2digit-1.png"></a> <a href="https://slowkow.github.io/hlabud/articles/examples.html#get-hla-allele-frequencies-from-allele-frequency-net-database-afnd"> <img width="49%" src="articles/examples_files/figure-html/afnd_dqb1_02_01-1.png"></a> <a href="https://slowkow.github.io/hlabud/articles/visualize-hla-structure.html"> <img width="49%" src="https://github.com/slowkow/ggrepel/assets/209714/4843a850-a4fd-4832-9600-0b8e9c1bb904"></a></p>
</div>
<div class="section level1">
<h1 id="citation">Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h1>
<p><code>hlabud</code> provides access to the data in IMGT/HLA database. Therefore, if you use <code>hlabud</code> then please cite the IMGT/HLA paper:</p>
<ul>
<li>Robinson J, Barker DJ, Georgiou X, Cooper MA, Flicek P, Marsh SGE. <a href="https://www.ncbi.nlm.nih.gov/pubmed/31667505" class="external-link">IPD-IMGT/HLA Database.</a> Nucleic Acids Res. 2020;48: D948–D955. <a href="doi:10.1093/nar/gkz950" class="uri">doi:10.1093/nar/gkz950</a>
</li>
</ul>
<p><code>hlabud</code> also provides access to the data in Allele Frequency Net Database (AFND). Therefore, if you use <code><a href="reference/hla_frequencies.html">hlabud::hla_frequencies()</a></code> then please cite the AFND paper:</p>
<ul>
<li>Gonzalez-Galarza FF, McCabe A, Santos EJMD, Jones J, Takeshita L, Ortega-Rivera ND, et al. <a href="https://pubmed.ncbi.nlm.nih.gov/31722398" class="external-link">Allele frequency net database (AFND) 2020 update: gold-standard data classification, open access genotype data and new query tools.</a> Nucleic Acids Res. 2020;48: D783–D788. <a href="doi:10.1093/nar/gkz1029" class="uri">doi:10.1093/nar/gkz1029</a>
</li>
</ul>
<p>Additionally, you can also cite the <code>hlabud</code> package like this:</p>
<ul>
<li>Slowikowski K. hlabud: methods for access and analysis of the human leukocyte antigen (HLA) gene sequence alignments from IMGT/HLA. R package version 1.0.0.</li>
</ul>
</div>
<div class="section level1">
<h1 id="related-work">Related work<a class="anchor" aria-label="anchor" href="#related-work"></a>
</h1>
<p>I recommend this article for anyone new to HLA, because the beautiful figures help to build intuition:</p>
<ul>
<li>La Gruta NL, Gras S, Daley SR, Thomas PG, Rossjohn J. <a href="https://www.ncbi.nlm.nih.gov/pubmed/29636542" class="external-link">Understanding the drivers of MHC restriction of T cell receptors.</a> Nat Rev Immunol. 2018;18: 467–478.</li>
</ul>
<p>Learn about the conventions for HLA nomenclature:</p>
<ul>
<li>Marsh SGE, Albert ED, Bodmer WF, Bontrop RE, Dupont B, Erlich HA, et al. <a href="https://www.ncbi.nlm.nih.gov/pubmed/20356336" class="external-link">Nomenclature for factors of the HLA system, 2010.</a> Tissue Antigens. 2010;75: 291–455.</li>
</ul>
<p>For case-control analysis of HLA genotype data, consider the <a href="https://CRAN.R-project.org/package=BIGDAWG" class="external-link">BIGDAWG</a> R package available on CRAN. Here is the related article:</p>
<ul>
<li>Pappas DJ, Marin W, Hollenbach JA, Mack SJ. <a href="https://pubmed.ncbi.nlm.nih.gov/26708359" class="external-link">Bridging ImmunoGenomic Data Analysis Workflow Gaps (BIGDAWG): An integrated case-control analysis pipeline.</a> Hum Immunol. 2016;77: 283–287.</li>
</ul>
</div>

  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/slowkow/hlabud/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/slowkow/hlabud/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small>GPL (&gt;= 3)</small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing hlabud</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Kamil Slowikowski <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-2843-6370" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Kamil Slowikowski.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
