<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Here are the conventions used for alignments (EBI IMGT-HLA):
The entry for each allele is displayed in respect to the reference sequences.
Where identity to the reference sequence is present the base will be displayed as a hyphen (-).
Non-identity to the reference sequence is shown by displaying the appropriate base at that position.
Where an insertion or deletion has occurred this will be represented by a period (.).
If the sequence is unknown at any point in the alignment, this will be represented by an asterisk (*).
In protein alignments for null alleles, the 'Stop' codons will be represented by a hash (X).
In protein alignments, sequence following the termination codon, will not be marked and will appear blank.
These conventions are used for both nucleotide and protein alignments.

"><title>Get aligned sequences in a dataframe — hla_alignments • hlabud</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Get aligned sequences in a dataframe — hla_alignments"><meta property="og:description" content="Here are the conventions used for alignments (EBI IMGT-HLA):
The entry for each allele is displayed in respect to the reference sequences.
Where identity to the reference sequence is present the base will be displayed as a hyphen (-).
Non-identity to the reference sequence is shown by displaying the appropriate base at that position.
Where an insertion or deletion has occurred this will be represented by a period (.).
If the sequence is unknown at any point in the alignment, this will be represented by an asterisk (*).
In protein alignments for null alleles, the 'Stop' codons will be represented by a hash (X).
In protein alignments, sequence following the termination codon, will not be marked and will appear blank.
These conventions are used for both nucleotide and protein alignments.

"><meta name="twitter:card" content="summary_large_image"><meta name="twitter:creator" content="@slowkow"><meta name="twitter:site" content="@slowkow"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">hlabud</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/examples.html">Examples</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/slowkow/hlabud">
    <span class="fa fa-github"></span>
     
    GitHub
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Get aligned sequences in a dataframe</h1>
      
      <div class="d-none name"><code>hla_alignments.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Here are the conventions used for alignments (<a href="https://www.ebi.ac.uk/ipd/imgt/hla/alignment/help/" class="external-link">EBI IMGT-HLA</a>):</p><ul><li><p>The entry for each allele is displayed in respect to the reference sequences.</p></li>
<li><p>Where identity to the reference sequence is present the base will be displayed as a hyphen (-).</p></li>
<li><p>Non-identity to the reference sequence is shown by displaying the appropriate base at that position.</p></li>
<li><p>Where an insertion or deletion has occurred this will be represented by a period (.).</p></li>
<li><p>If the sequence is unknown at any point in the alignment, this will be represented by an asterisk (*).</p></li>
<li><p>In protein alignments for null alleles, the 'Stop' codons will be represented by a hash (X).</p></li>
<li><p>In protein alignments, sequence following the termination codon, will not be marked and will appear blank.</p></li>
<li><p>These conventions are used for both nucleotide and protein alignments.</p></li>
</ul></div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">hla_alignments</span><span class="op">(</span>gene <span class="op">=</span> <span class="st">"DRB1"</span>, type <span class="op">=</span> <span class="st">"prot"</span>, release <span class="op">=</span> <span class="cn">NULL</span>, quiet <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>gene</dt>
<dd><p>The name of a gene like "DRB1"</p></dd>


<dt>type</dt>
<dd><p>The type of sequence, one of "prot", "nuc", "gen"</p></dd>


<dt>release</dt>
<dd><p>The name of a release like "3.51.0" (optional)</p></dd>


<dt>quiet</dt>
<dd><p>If FALSE, print messages along the way.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A dataframe.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># \donttest{</span></span></span>
<span class="r-in"><span><span class="va">a</span> <span class="op">&lt;-</span> <span class="fu">hla_alignments</span><span class="op">(</span><span class="st">"DRB1"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">a</span><span class="op">$</span><span class="va">sequences</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             allele</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1 DRB1*01:01:01:01</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2 DRB1*01:01:01:02</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 DRB1*01:01:01:03</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4 DRB1*01:01:01:04</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5 DRB1*01:01:01:05</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6 DRB1*01:01:01:06</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                                                                                                                                                                                                                                                             seq</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1 MVCLKLPGGSCMTALTVTLMVLSSPLALAGDTRPRFLWQLKFECHFFNGTERVR.LLERCIYNQEE.SVRFDSDVGEYRAVTELGRPDAEYWNSQKDLLEQRRAAVDTYCRHNYGVGESFTVQRR.VEPKVTVYPSKTQPLQHHNLLVCSVSGFYPGSIEVRWFRNGQEEKAGVVSTGLIQNGDWTFQTLVMLETVPRSGEVYTCQVEHPSVTSPLTVEWRARSESAQSKMLSGVGGFVLGLLFLGAGLFIYFRNQKGHSGLQPTGFLS</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2 ------------------------------------------------------.-----------.----------------------------------------------------------.-----------------------------------------------------------------------------------------------------------------------------------------------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 ------------------------------------------------------.-----------.----------------------------------------------------------.-----------------------------------------------------------------------------------------------------------------------------------------------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4 ------------------------------------------------------.-----------.----------------------------------------------------------.-----------------------------------------------------------------------------------------------------------------------------------------------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5 ------------------------------------------------------.-----------.----------------------------------------------------------.-----------------------------------------------------------------------------------------------------------------------------------------------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6 ------------------------------------------------------.-----------.----------------------------------------------------------.-----------------------------------------------------------------------------------------------------------------------------------------------</span>
<span class="r-in"><span><span class="va">a</span><span class="op">$</span><span class="va">aminos</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                  Pn29 Pn28 Pn27 Pn26 Pn25 Pn24</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DRB1*01:01:01:01    M    V    C    L    K    L</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DRB1*01:01:01:02    M    V    C    L    K    L</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DRB1*01:01:01:03    M    V    C    L    K    L</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DRB1*01:01:01:04    M    V    C    L    K    L</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DRB1*01:01:01:05    M    V    C    L    K    L</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DRB1*01:01:01:06    M    V    C    L    K    L</span>
<span class="r-in"><span><span class="va">a</span><span class="op">$</span><span class="va">onehot</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                  Pn29_M Pn28_L Pn28_V Pn27_C Pn26_L Pn25_K</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DRB1*01:01:01:01      1      0      1      1      1      1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DRB1*01:01:01:02      1      0      1      1      1      1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DRB1*01:01:01:03      1      0      1      1      1      1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DRB1*01:01:01:04      1      0      1      1      1      1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DRB1*01:01:01:05      1      0      1      1      1      1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DRB1*01:01:01:06      1      0      1      1      1      1</span>
<span class="r-in"><span><span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Kamil Slowikowski.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

