---
title: "hlabud usage examples"
vignette: >
  %\VignetteIndexEntry{hlabud usage examples}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
#output:
#  prettydoc::html_pretty:
#    css: custom.css
#    theme: hpstr
#    highlight: github
#    toc: true
#    keep_md: true
options(width=80)

library(data.table)
library(dplyr)
library(parameters)
library(pbapply)
library(glue)
library(scales)
library(readr)
library(stringr)
library(magrittr)
library(ggplot2)
library(ggrepel)
library(ggstance)
file_size <- function(x) glue("{fs::file_size(x)}B")

#' @importFrom ggplot2 theme_classic theme element_rect element_line
#'   element_blank element_text
#' @importFrom grid unit
theme_kamil <- theme_classic(
  base_family = "Arial",
  base_size = 16,
  base_line_size = 0.3,
  base_rect_size = 0.3
) +
theme(
  panel.spacing    = unit(2, "lines"),
  panel.border     = element_rect(linewidth = 0.5, fill = NA),
  axis.ticks       = element_line(linewidth = 0.4),
  axis.line        = element_blank(),
  strip.background = element_blank(),
  plot.title       = element_text(size = 16),
  plot.title.position = "plot",
  plot.subtitle    = element_text(size = 16),
  plot.caption     = element_text(size = 16),
  strip.text       = element_text(size = 16),
  legend.text      = element_text(size = 16),
  legend.title     = element_text(size = 16),
  axis.text        = element_text(size = 16),
  axis.title       = element_text(size = 16)
)
theme_set(theme_kamil)
options(
  ggplot2.discrete.colour = c(
    "#666666", "#E69F00", "#56B4E9", "#009E73", "#0072B2", "#D55E00", "#CC79A7"
  ),
  ggplot2.discrete.fill = c(
    "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7", "#666666"
  )
)

mpn65 <- c(
  '#ff0029', '#377eb8', '#66a61e', '#984ea3', '#00d2d5', '#ff7f00', '#af8d00',
  '#7f80cd', '#b3e900', '#c42e60', '#a65628', '#f781bf', '#8dd3c7', '#bebada',
  '#fb8072', '#80b1d3', '#fdb462', '#fccde5', '#bc80bd', '#ffed6f', '#c4eaff',
  '#cf8c00', '#1b9e77', '#d95f02', '#e7298a', '#e6ab02', '#a6761d', '#0097ff',
  '#00d067', '#000000', '#252525', '#525252', '#737373', '#969696', '#bdbdbd',
  '#f43600', '#4ba93b', '#5779bb', '#927acc', '#97ee3f', '#bf3947', '#9f5b00',
  '#f48758', '#8caed6', '#f2b94f', '#eff26e', '#e43872', '#d9b100', '#9d7a00',
  '#698cff', '#d9d9d9', '#00d27e', '#d06800', '#009f82', '#c49200', '#cbe8ff',
  '#fecddf', '#c27eb6', '#8cd2ce', '#c4b8d9', '#f883b0', '#a49100', '#f48800',
  '#27d0df', '#a04a9b'
)

knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)

library(hlabud)
my_gene <- "DRB1"
a <- hla_alignments(gene = my_gene, type = "prot")
n_alleles <- nrow(a$seq)

```

# Introduction

Kamil Slowikowski

`r Sys.Date()`

[hlabud] is an R package that provides functions to facilitate download and analysis of human leukocyte antigen (HLA) genotype sequence alignments from [IMGTHLA] in R.

[hlabud]: https://github.com/slowkow/hlabud
[IMGTHLA]: https://github.com/ANHIG/IMGTHLA

Let's consider a question that we might want to answer about HLA genotypes.
What amino acid positions are different between the DRB1\*04:174 and DRB1\*15:152 genotypes?

With hlabud, we can find the answer with a few lines of code:

```{r drb1_dosage_amino}
library(hlabud)
a <- hla_alignments("DRB1")
dosage(c("DRB1*04:174", "DRB1*15:152"), a$onehot)
```

The two genotypes are nearly identical, but the amino acid at position 9 is different:

- position 9 E (Glu) in DRB1\*04:174
- position 9 W (Trp) in DRB1\*15:152

It is just as easy to find the nucleotides that distinguish the two alleles:

```{r drb1_dosage_nucleotide}
n <- hla_alignments("DRB1", type = "nuc")
dosage(c("DRB1*04:174", "DRB1*15:152"), n$onehot)
```

# Installation

The quickest way to get hlabud is to install from GitHub:

```{r install, eval=FALSE}
# install.packages("devtools")
devtools::install_github("slowkow/hlabud")
```

Below, I included a few usage examples. I hope they inspire you to share your own HLA analyses.

The source code for this page is available [here](https://github.com/slowkow/hlabud/tree/main/vignettes/examples.Rmd).

Thank you for [reporting issues](https://github.com/slowkow/hlabud/issues) with hlabud.


# Get a one-hot encoded matrix for all HLA-`r my_gene` alleles

We can use `hla_alignments("DRB1")` to load the `DRB1_prot.txt` file from the
latest [IMGTHLA](https://github.com/ANHIG/IMGTHLA/releases) release:

```{r}
library(hlabud)
a <- hla_alignments(gene = "DRB1", verbose = TRUE)
```

The `a` object is a list with three items:

```{r}
str(a)
```

`a$sequences` has amino acid sequence alignments in a data frame:

```{r}
a$sequences
```

Here are the conventions used for alignments (copied from EBI):

> * The entry for each allele is displayed in respect to the reference sequences.
> * Where identity to the reference sequence is present the base will be displayed as a hyphen (-).
> * Non-identity to the reference sequence is shown by displaying the appropriate base at that position.
> * Where an insertion or deletion has occurred this will be represented by a period (.).
> * If the sequence is unknown at any point in the alignment, this will be represented by an asterisk (*).
> * In protein alignments for null alleles, the 'Stop' codons will be represented by a hash (X).
> * In protein alignments, sequence following the termination codon, will not be marked and will appear blank.
> * These conventions are used for both nucleotide and protein alignments.

Learn more about the alignments at: 

* https://www.ebi.ac.uk/ipd/imgt/hla/alignment/help/

`a$alleles` has a matrix of amino acids with one column for each position:

```{r}
a$alleles[1:5,1:7]
```

`a$onehot` has a one-hot encoded matrix with one column for each amino acid at
each position:

```{r}
a$onehot[1:5,1:7]
```

# Convert genotypes to a dosage matrix

Suppose we have some individuals with the following genotypes:

```{r, eval=FALSE, include=FALSE}
dput(paste(sample(rownames(a$onehot), 5), sample(rownames(a$onehot), 5), sep = "+"))
```

```{r}
genotypes <- c(
  "DRB1*12:02:02:03,DRB1*12:02:02:03,DRB1*14:54:02",
  "DRB1*04:174,DRB1*15:152",
  "DRB1*04:56:02,DRB1*15:01:48",
  "DRB1*14:172,DRB1*04:160",
  "DRB1*04:359,DRB1*04:284:02"
)
```

If we want to run an association test on the amino acid positions, then we need
to convert the genotype names to a matrix of allele dosages (e.g., 0, 1, 2).

We can use `dosage()` to convert each individual's genotypes to amino
acid dosages:

```{r}
dosage <- dosage(genotypes, a$onehot)
dosage[,1:4]
dim(dosage)
```

**Note:**

* The `dosage` matrix has one row for each individual and one column for each
  amino acid at each position. By default, `dosage()` will discard the
  columns where all individuals are identical.

* The first individual has `dosage=3` for `pos6_R` (position 6 Arg). That's
  because we assigned this individual 3 alleles in our input.

Please be careful to check that your data looks the way you expect!

# Logistic regression association for amino acid positions

Let's simulate a dataset with cases and controls to demonstrate one approach
for testing which amino acid positions might be associated with cases.

```{r}
set.seed(2)
n <- 100
d <- data.frame(
  geno = paste(
    sample(rownames(a$onehot), n, replace = TRUE),
    sample(rownames(a$onehot), n, replace = TRUE),
    sep = ","
  ),
  age = sample(21:100, n, replace = TRUE),
  case = sample(0:1, n, replace = TRUE)
)
d <- cbind(d, dosage(d$geno, a$onehot))
d[1:5,1:6]
```

Our simulated dataset has `r n` individuals, `r sum(d$case == 1)` cases and `r sum(d$case == 0)` controls. We also have one column for each amino acid position that we might want to test for association with the `case` variable.

One possible approach for association testing is to use `glm()` to fit a
logistic regression model for each amino acid position. This could reveal if
any amino acid position might be associated with the `case` variable in our
simulated dataset.

```{r glm, cache=TRUE, warning=FALSE, message=FALSE}

# select the amino acid positions that have at least 3 people with dosage > 0
my_as <- names(which(colSums(d[,4:ncol(d)] > 0) >= 3))

# run the association tests
my_glm <- rbindlist(pblapply(my_as, function(my_a) {
  f <- sprintf("case ~ %s", my_a)
  glm(as.formula(f), data = d, family = "binomial") %>%
    parameters(exponentiate = TRUE)
}))

# look at the top hits
my_glm %>%
  arrange(p) %>%
  filter(!Parameter %in% c("(Intercept)")) %>%
  head
```

The volcano below shows the Odds Ratio and P-value for each amino acid
position. The top hits with P &lt; 0.05 are labeled.

```{r glm-volcano, fig.width = 6, fig.height = 4, fig.dpi = 300, echo = FALSE}

my_glm_case <- my_glm %>%
  filter(!Parameter %in% c("(Intercept)")) %>%
  filter(Coefficient < 10) %>%
  arrange(p)

p <- ggplot(my_glm_case) +
  aes(Coefficient, -log10(p)) +
  geom_vline(xintercept = 1, linewidth = 0.3, alpha = 0.3) +
  geom_point() +
  geom_text_repel(
    data = \(d) d %>% filter(p < 0.05),
    mapping = aes(label = Parameter)
  ) +
  labs(
    x = "Odds Ratio",
    y = "-log10 P",
    title = "Association between amino acid positions and case status",
    subtitle = glue("{nrow(my_glm_case)} amino acid positions")
  )
p
# ggsave("test.png", p)

```

In this simulation, the `case` variable is associated with `r my_glm_case$Parameter[1]` `r with(my_glm_case[1,], glue("(P = {signif(p, 2)}, OR = {signif(Coefficient, 2)}, 95% CI {signif(CI_low, 2)} to {signif(CI_high, 2)})"))`.


# UMAP embedding of `r comma(n_alleles)` HLA-`r my_gene` alleles

There are many possibilities for analysis of a one-hot encoding of HLA-`r my_gene` alleles.

For example, here is a UMAP embedding of `r comma(n_alleles)` HLA-`r my_gene` alleles encoded as a one-hot amino acid matrix with `r ncol(a$onehot)` columns, one for each amino acid at each position.

```{r, eval=FALSE}
uamp(a$onehot, n_epochs = 200, min_dist = 1, spread = 2)
```

```{r umap-2digit, cache=TRUE, echo = FALSE, fig.dpi = 300, fig.width = 9, fig.height = 6, message=FALSE, warning=FALSE}

library(uwot)

set.seed(2)
d <- umap(a$onehot, n_epochs = 200, min_dist = 1, spread = 2)

a$umap <- data.frame(allele = rownames(a$onehot), x = d[,1], y = d[,2])
a$umap %<>% mutate(allele = str_remove(allele, ".+\\*"))
a$umap %<>% mutate(d4 = str_extract(allele, "^[^:]+:[^:]+"))
a$umap %<>% mutate(d2 = str_extract(d4, "^[^:]+"))
a$umap %<>% mutate(m2 = str_extract(d4, "[^:]+$"))
a$umap$label <- ""
set.seed(2)
ix <- sample(1:nrow(a$umap), 40)
a$umap$label[ix] <- a$umap$d4[ix]

my_release <- hla_releases()[1]

p <- ggplot(a$umap) +
  aes(x, y, color = d2) +
  geom_point() +
  geom_text_repel(aes(label = label), max.overlaps = 100) +
  scale_color_manual(values = mpn65) + 
  guides(color = guide_legend(title = "2-digit", override.aes = list(label = "", size = 5))) +
  labs(
    title = glue("{comma(nrow(a$umap))} HLA-{my_gene} alleles"),
    caption = glue("IMGTHLA {my_release}")
  ) +
  theme(
    axis.text = element_blank(),
    axis.ticks = element_blank(),
    axis.title = element_blank(),
  )
p
# ggsave("test.png", p)

```

We can highlight which alleles have amino acid H at position 13:

```{r umap-pos13_H, cache=TRUE, echo = FALSE, fig.dpi = 300, fig.width = 9, fig.height = 6, message=FALSE, warning=FALSE}

my_position <- "pos13_H"
a$umap$color <- as.factor(as.integer(a$onehot[,my_position]))
p <- ggplot(a$umap) +
  aes(x, y, color = color) +
  geom_point() +
  geom_text_repel(aes(label = label), max.overlaps = 100) +
  guides(color = guide_legend(title = my_position, override.aes = list(label = "", size = 5))) +
  labs(
    title = glue("{comma(nrow(a$umap))} HLA-{my_gene} alleles"),
    caption = glue("IMGTHLA {my_release}")
  ) +
  theme(
    axis.text = element_blank(),
    axis.ticks = element_blank(),
    axis.title = element_blank(),
  )
p
# ggsave("test.png", p)

```

Or we can represent each amino acid at position 13 with a different color:

```{r umap-pos13, cache=TRUE, echo = FALSE, fig.dpi = 300, fig.width = 9, fig.height = 6, message=FALSE, warning=FALSE}

my_position <- "pos13"
a$umap$color <- as.character(a$alleles[,my_position])

one_to_three <- c(
  "A" = "Ala", "C" = "Cys", "D" = "Asp", "E" = "Glu", "F" = "Phe", "G" = "Gly",
  "H" = "His", "I" = "Ile", "K" = "Lys", "L" = "Leu", "M" = "Met", "N" = "Asn",
  "P" = "Pro", "Q" = "Gln", "R" = "Arg", "S" = "Ser", "T" = "Thr", "V" = "Val",
  "W" = "Trp", "Y" = "Tyr", "*" = "Ter"
)
aa_count <- a$umap %>% count(color) %>% arrange(-n) %>% mutate(label = glue("{one_to_three[color]} (n={n})"))
aa_to_count <- aa_count$label
names(aa_to_count) <- aa_count$color
a$umap$color <- factor(a$umap$color, aa_count$color)

p <- ggplot(a$umap) +
  aes(x, y, color = color) +
  geom_point() +
  geom_text_repel(aes(label = label), max.overlaps = 100) +
  guides(color = guide_legend(title = my_position, override.aes = list(label = "", size = 5))) +
  scale_color_manual(values = mpn65, labels = aa_to_count) +
  labs(
    title = glue("{comma(nrow(a$umap))} HLA-{my_gene} alleles"),
    caption = glue("IMGTHLA {my_release}")
  ) +
  theme(
    axis.text = element_blank(),
    axis.ticks = element_blank(),
    axis.title = element_blank(),
  )
p
# ggsave("test.png", p)

```

# Get HLA allele frequencies from Allele Frequency Net Database (AFND)

Download and read a table of HLA allele frequencies from the [Allele Frequency Net Database (AFND)](http://www.allelefrequencies.net/).

If you use this data, please cite the latest manuscript about Allele Frequency Net Database:

- Gonzalez-Galarza FF, McCabe A, Santos EJMD, Jones J, Takeshita L, Ortega-Rivera ND, et al. [Allele frequency net database (AFND) 2020 update: gold-standard data classification, open access genotype data and new query tools.](https://pubmed.ncbi.nlm.nih.gov/31722398) Nucleic Acids Res. 2020;48: D783–D788. doi:10.1093/nar/gkz1029

```{r afnd}
af <- hla_frequencies()
af
```

Plot the frequency of a specific allele (DQB1*02:01) in populations with more than 1000 sampled individuals:

```{r afnd_dqb1_02_01, fig.dpi = 300, fig.width = 10, fig.height = 8}
my_allele <- "DQB1*02:01"
my_af <- af %>% filter(allele == my_allele) %>%
  filter(n > 1000) %>%
  arrange(-alleles_over_2n)

ggplot(my_af) +
  aes(x = alleles_over_2n, y = reorder(population, alleles_over_2n)) +
  scale_y_discrete(position = "right") +
  geom_colh() +
  labs(
    x = "Allele Frequency (Alleles / 2N)",
    y = NULL,
    title =  glue("Frequency of {my_allele} across {length(unique(my_af$population))} populations"),
    caption = "Data from AFND http://allelefrequencies.net"
  )
```

See [github.com/slowkow/allelefrequencies](https://github.com/slowkow/allelefrequencies) for more examples of how we might use this data.


# Compute HLA divergence with the Grantham distance matrix

Each HLA allele binds a specific set of peptides. So, an individual with two highly dissimilar alleles can bind a greater number of different peptides than a homozygous individual (https://doi.org/10.1007/BF02918202):

> Each MHC class II allele has the capacity to bind and present a specific set of peptides from processed antigens. The inability of a specific class II allele to bind and present a fragment derived from a processed antigen results in the loss of immune responsiveness for that antigen in individuals homozygous for that class II allele.

We can compute an HLA divergence metric for a set of individuals like this:

```{r div-grantham}
my_genos <- c("A*23:01:12,A*24:550", "A*25:12N,A*11:27", "A*24:381,A*33:85")

hla_divergence(my_genos, method = "grantham")
```

The divergence for a homozygote is equal to zero, by definition:

```{r div-homozygote}
hla_divergence("A*01:01,A*01:01")
```

By default, we use the amino acid distance matrix from Granthan 1974 (https://doi.org/10.1126/science.185.4154.862).

Alternatively, we can choose to use a uniform matrix instead (diagonal values are 0, all non-diagonal values are equal to 1):

```{r div-uniform}
hla_divergence(my_genos, method = "uniform")
```

The amino acid distance matrix is easily accessible, and we provide two built-in options `"grantham"` and `"uniform"`:

```{r amino-dist}
amino_distance_matrix(method = "uniform")
```


# Download and unpack all data from the latest IMGTHLA release

If you only want to use `hla_alignments()`, then you don't need `install_hla()` because data files are downloaded automatically as needed and cached for future use.

But some users might need access to additional files that are only present in the full data release.

Run `install_hla()` to download and unpack the [latest IMGTHLA release](https://github.com/ANHIG/IMGTHLA/releases).
The destination folder for the downloaded data files is `getOption("hlabud_dir")` (automatically tailored to your operating system thanks to the [rappdirs](https://github.com/r-lib/rappdirs) package).

Here are a few examples of how to download releases or get a list of release names.

Download the latest release (default) or a specific release:

```{r install_hla, eval=FALSE}
# Download all of the data (120MB) for the latest IMGTHLA release
install_hla(release = "latest")

# Download a specific release
install_hla(release = "3.51.0")
```

Optionally, get or set the directory hlabud uses to store the data:

```{r hlabud_dir, eval=FALSE}
getOption("hlabud_dir")
#> [1] "/home/username/.local/share/hlabud"

# Manually override the directory for hlabud to use
options(hlabud_dir = "/path/to/my/dir")
```

List all releases:

```{r hla_releases, eval=FALSE}
hla_releases()
#>  [1] "3.51.0"   "3.50.0"   "3.49.0"   "3.48.0"   "3.47.0"   "3.46.0"   "3.45.1"   "3.45.01"
#>  [9] "3.45.0.1" "3.45.0"   "3.44.1"   "3.44.0"   "3.43.0"   "3.42.0"   "3.41.2"   "3.41.0"
#> [17] "3.40.0"   "3.39.0"   "3.38.0"   "3.37.0"   "3.36.0"   "3.35.0"   "3.34.0"   "3.33.0"
#> [25] "3.32.0"   "3.31.0"   "3.30.0"   "3.29.0"   "3.28.0"   "3.27.0"
```

After installing a few releases, the hlabud folder might look like this:

```
❯ ls -lah "/home/user/.local/share/hlabud"
total 207M
drwxrwxr-x  3 user user      32 Apr  5 01:19 3.30.0
drwxrwxr-x 11 user user    4.0K Apr  7 19:31 3.40.0
drwxrwxr-x 12 user user    4.0K Apr  5 00:27 3.51.0
-rw-rw-r--  1 user user     15K Apr  7 19:23 tags.json
-rw-rw-r--  1 user user     79M Apr  7 19:28 v3.40.0-alpha.tar.gz
-rw-rw-r--  1 user user    129M Apr  4 20:07 v3.51.0-alpha.tar.gz
```

# Count the number of alleles in each IMGTHLA release

We can get a list of the release names:

```{r releases}
releases <- hla_releases()
releases
```

Then we can get the allele names for each release:

```{r hla_alleles}
my_alleles <- rbindlist(lapply(releases, function(release) {
  retval <- hla_alleles(release = release)
  retval$release <- release
  return(retval)
}), fill = TRUE)
```

Next, count how many alleles we have in each release:

```{r alleles_per_release_count}
d <- my_alleles %>% count(release) %>% filter(n > 1)
d
```

And plot the number of alleles as a line plot:

```{r alleles_per_release, fig.height = 6}
ggplot(d) +
  aes(x = release, y = n, group = 1) +
  geom_line() +
  geom_text(aes(label = release), hjust = 1) +
  labs(x = NULL, y = "Number of alleles",
  title = "Each release has more HLA alleles") +
  theme(
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
  )
```

```{r alleles_per_release_and_gene, fig.height = 12}
d2 <- my_alleles %>% mutate(gene = str_split_fixed(Allele, "\\*", 2)[,1]) %>% count(release, gene)
ggplot() +
  aes(x = release, y = n) +
  geom_line(
    data = d2,
    aes(group = gene, color = gene)
  ) +
  scale_color_discrete(guide = "none") +
  geom_text(
    data = d2 %>% filter(release == "3.52.0"),
    mapping = aes(label = gene),
    hjust = 0
  ) +
  labs(x = NULL, y = "Number of alleles",
  title = "Number of alleles per release and gene") +
  scale_x_discrete(expand = expansion(mult = c(0.01, 0.1))) +
  scale_y_log10() +
  theme(
    panel.grid.major.y = element_line(), 
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
  )
```
