---
title: "Visualize HLA protein structures"
vignette: >
  %\VignetteIndexEntry{Visualize HLA protein structures}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Introduction

Kamil Slowikowski

`r Sys.Date()`

In this vignette, we explore a few different methods for visualizing the molecular structure of HLA proteins.
First, we'll look at an example of how to use the [NGLVieweR] R package to show HLA protein structures.
Next, we'll use [PyMOL] to do the same thing.

# What are the PDB identifiers for each HLA gene?

Here is a list of PDB identifiers you might consider using to represent each HLA protein:

```
HLA-A  2xpg
HLA-B  2bvp
HLA-C  4nt6
HLA-DP 3lqz
HLA-DQ 4z7w
HLA-DR 3pdo
```

Also try searching the PDB website for, e.g., `"HLA-DR"` and see if there is a more appropriate structure for your analysis.

# NGLVieweR

Let's try to visualize position 9 in the HLA-B protein structure.

We will visualize the structure of [2bvp] from the Protein Data Bank (PDB).

Here is an example of how to do this with the [NGLVieweR] R package by [Niels van der Velden]:

[2bvp]: https://www.rcsb.org/structure/2BVP
[NGLVieweR]: https://github.com/nvelden/NGLVieweR
[Niels van der Velden]: https://www.nielsvandervelden.com/

```{r, nglviewer-2bvp}
# devtools::install_github("nvelden/NGLVieweR") # we need the latest version
library(NGLVieweR)
library(magrittr)
my_sele <- "9:A"
NGLVieweR("2bvp") %>%
  stageParameters(
    backgroundColor = "white",
    zoomSpeed = 1,
    cameraFov = 80
  ) %>%
  addRepresentation(
    type = "cartoon"
  ) %>%
  addRepresentation(
    type = "ball+stick",
    param = list(
      sele = my_sele
    )
  ) %>%
  addRepresentation(
    type = "label",
    param = list(
      sele = my_sele,
      labelType = "format",
      labelFormat = "[%(resname)s]%(resno)s", # or enter custom text
      labelGrouping = "residue", # or "atom" (eg. sele = "20:A.CB")
      color = "black",
      fontFamiliy = "sans-serif",
      xOffset = 1,
      yOffset = 0,
      zOffset = 0,
      fixedSize = TRUE,
      radiusType = 1,
      radiusSize = 5.5, # Label size
      showBackground = TRUE
      # backgroundColor="black",
      # backgroundOpacity=0.5
    )
  ) %>%
  zoomMove(
    center = my_sele,
    zoom = my_sele,
    duration = 0, # animation time in ms
    z_offSet = -20
  ) %>%
  setSpin()
```


In the view above, we see the blue peptide and the red HLA-B protein.
The tyrosine at position 9 is highlighted with a `ball+stick` representation, and it is also labeled with a text label.
The structure is rotating so we can getter a better view.

We can use hlabud to answer some questions about HLA-B Tyr9 (tyrosine at position 9).

For example, which HLA-B alleles have this amino acid position?

```{r, B-Tyr9}
library(hlabud)
a <- hla_alignments("B")
head(names(which(a$onehot[,"pos9_Y"] == 1)))
```

What fraction of reported HLA-B alleles have Tyr9?

```{r, B-Tyr9-sum}
sum(a$onehot[,"pos9_Y"] == 1) / nrow(a$onehot)
```

# PyMOL

[PyMOL] is one of my favorite methods for visualizing protein structures, because it allows us to [change a residue][mutagenesis] in an existing protein and visualize the new mutated protein.

It only takes few lines of PyMOL to create a nice figure.

For example, if we want to quickly highlight positions 13 and 45 in HLA-DQB1, this snippet of PyMOL code will produce the figure below.

Here is a Bash script that will:

1. Write a PyMOL script
2. Run the PyMOL script with the `pymol` command

```{bash, eval=FALSE}
#!/usr/bin/env bash

# Write a pymol script
cat << EOF > script.pml
fetch 7kei
show cartoon
color teal, chain A
color orange, chain B
color purple, chain C
color red, chain B & resi 13
color red, chain B & resi 45
label n. CA and chain B & resi 13, "%s %s" % (resi, resn)
label n. CA and chain B & resi 45, "%s %s" % (resi, resn)
png 7kei.png, width=1200, height=800, dpi=300 
EOF

# On Linux, we can just use `pymol` without making an alias

# On macOS, we need to make an alias
alias pymol=/Applications/PyMOL.app/Contents/MacOS/PyMOL

pymol -c script.pml
```

Here is what the PyMOL script will do:

1. Load a structure from the [Protein Data Bank (PDB)](https://www.rcsb.org/). `7kei` is the identifier for [a published protein structure](https://www.rcsb.org/structure/7KEI).
2. Color the HLA-DQA1 protein teal.
3. Color the HLA-DQB1 protein orange.
4. Color the peptide purple.
5. We color residues 13 and 45 in HLA-DQB1 red.
6. Label those residues with their positions and names.
7. Write a PNG file with a view of the structure.

<img src="https://github.com/slowkow/ggrepel/assets/209714/4843a850-a4fd-4832-9600-0b8e9c1bb904"></img>

In the image above, I manually rotated the structure with my mouse and added more text labels like `"PDB: 7kei"` after saving the file.

# Other PDB viewers

Python:

* https://github.com/nglviewer/nglview

Javascript:

* https://www.rcsb.org/3d-view
* https://www.ncbi.nlm.nih.gov/Structure/icn3d/full.html?mmdbid=7kei&bu=1
* https://github.com/nglviewer/ngl
* https://github.com/biasmv/pv

R:

* https://www.raymolecule.com

[mutagenesis]: https://pymolwiki.org/index.php/Mutagenesis

