<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="hlabud">
<title>hlabud usage examples • hlabud</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="hlabud usage examples">
<meta property="og:description" content="hlabud">
<meta name="twitter:card" content="summary_large_image">
<meta name="twitter:creator" content="@slowkow">
<meta name="twitter:site" content="@slowkow">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">hlabud</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/examples.html">Examples</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/slowkow/hlabud">
    <span class="fa fa-github"></span>
     
    GitHub
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>hlabud usage examples</h1>
            
      
      
      <div class="d-none name"><code>examples.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Kamil Slowikowski</p>
<p>2023-06-12</p>
<p><a href="https://github.com/slowkow/hlabud" class="external-link">hlabud</a> is an R
package that provides functions to download and analyze human leukocyte
antigen (HLA) genotypes from <a href="https://github.com/ANHIG/IMGTHLA" class="external-link">IMGTHLA</a> in a tidy R
workflow.</p>
<p>For example, what amino acid positions are different between the
DRB1*04:174 and DRB1*15:152 genotypes?</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">hlabud</span><span class="op">)</span></span>
<span><span class="va">a</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hla_alignments.html">hla_alignments</a></span><span class="op">(</span><span class="st">"DRB1"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/amino_dosage.html">amino_dosage</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"DRB1*04:174"</span>, <span class="st">"DRB1*15:152"</span><span class="op">)</span>, <span class="va">a</span><span class="op">$</span><span class="va">onehot</span><span class="op">)</span></span>
<span><span class="co">#&gt;             P9_E P9_W</span></span>
<span><span class="co">#&gt; DRB1*04:174    1    0</span></span>
<span><span class="co">#&gt; DRB1*15:152    0    1</span></span></code></pre></div>
<p>The two genotypes are nearly identical, but the amino acid at
position 9 is different:</p>
<ul>
<li>E (Glu) in DRB1*04:174</li>
<li>W (Trp) in DRB1*15:152</li>
</ul>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>The quickest way to get hlabud is to install from GitHub:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install.packages("devtools")</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"slowkow/hlabud"</span><span class="op">)</span></span></code></pre></div>
<p>Below, I included a few usage examples. I hope they inspire you to
share your own HLA analyses.</p>
<p>The source code for this page is available <a href="https://github.com/slowkow/hlabud/tree/main/vignettes/examples.Rmd" class="external-link">here</a>.</p>
<p>Thank you for <a href="https://github.com/slowkow/hlabud/issues" class="external-link">reporting issues</a>
with hlabud.</p>
</div>
<div class="section level2">
<h2 id="get-a-one-hot-encoded-matrix-for-all-hla-drb1-alleles">Get a one-hot encoded matrix for all HLA-DRB1 alleles<a class="anchor" aria-label="anchor" href="#get-a-one-hot-encoded-matrix-for-all-hla-drb1-alleles"></a>
</h2>
<p>We can use <code>hla_alignments("DRB1")</code> to load the
<code>DRB1_prot.txt</code> file from the latest <a href="https://github.com/ANHIG/IMGTHLA/releases" class="external-link">IMGTHLA</a>
release:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">hlabud</span><span class="op">)</span></span>
<span><span class="va">a</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hla_alignments.html">hla_alignments</a></span><span class="op">(</span>gene <span class="op">=</span> <span class="st">"DRB1"</span>, quiet <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt; hlabud is using IMGTHLA release 3.52.0</span></span>
<span><span class="co">#&gt; Reading /home/runner/.local/share/hlabud/3.52.0/alignments/DRB1_prot.txt</span></span></code></pre></div>
<p>The <code>a</code> object is a list with three items:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">a</span>, max.level <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 3</span></span>
<span><span class="co">#&gt;  $ sequences:'data.frame':   3516 obs. of  2 variables:</span></span>
<span><span class="co">#&gt;  $ aminos   :'data.frame':   3516 obs. of  288 variables:</span></span>
<span><span class="co">#&gt;  $ onehot   : num [1:3516, 1:1482] 1 1 1 1 1 1 1 1 1 1 ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "dimnames")=List of 2</span></span></code></pre></div>
<p><code>a$sequences</code> has amino acid sequence alignments in a data
frame:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">a</span><span class="op">$</span><span class="va">sequences</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="op">]</span></span>
<span><span class="co">#&gt;             allele</span></span>
<span><span class="co">#&gt; 1 DRB1*01:01:01:01</span></span>
<span><span class="co">#&gt; 2 DRB1*01:01:01:02</span></span>
<span><span class="co">#&gt; 3 DRB1*01:01:01:03</span></span>
<span><span class="co">#&gt; 4 DRB1*01:01:01:04</span></span>
<span><span class="co">#&gt; 5 DRB1*01:01:01:05</span></span>
<span><span class="co">#&gt;                                                                                                                                                                                                                                                                             seq</span></span>
<span><span class="co">#&gt; 1 MVCLKLPGGSCMTALTVTLMVLSSPLALAGDTRPRFLWQLKFECHFFNGTERVR.LLERCIYNQEE.SVRFDSDVGEYRAVTELGRPDAEYWNSQKDLLEQRRAAVDTYCRHNYGVGESFTVQRR.VEPKVTVYPSKTQPLQHHNLLVCSVSGFYPGSIEVRWFRNGQEEKAGVVSTGLIQNGDWTFQTLVMLETVPRSGEVYTCQVEHPSVTSPLTVEWRARSESAQSKMLSGVGGFVLGLLFLGAGLFIYFRNQKGHSGLQPTGFLS</span></span>
<span><span class="co">#&gt; 2 ------------------------------------------------------.-----------.----------------------------------------------------------.-----------------------------------------------------------------------------------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; 3 ------------------------------------------------------.-----------.----------------------------------------------------------.-----------------------------------------------------------------------------------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; 4 ------------------------------------------------------.-----------.----------------------------------------------------------.-----------------------------------------------------------------------------------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; 5 ------------------------------------------------------.-----------.----------------------------------------------------------.-----------------------------------------------------------------------------------------------------------------------------------------------</span></span></code></pre></div>
<p>Here are the conventions used for alignments (copied from EBI):</p>
<blockquote>
<ul>
<li>The entry for each allele is displayed in respect to the reference
sequences.</li>
<li>Where identity to the reference sequence is present the base will be
displayed as a hyphen (-).</li>
<li>Non-identity to the reference sequence is shown by displaying the
appropriate base at that position.</li>
<li>Where an insertion or deletion has occurred this will be represented
by a period (.).</li>
<li>If the sequence is unknown at any point in the alignment, this will
be represented by an asterisk (*).</li>
<li>In protein alignments for null alleles, the ‘Stop’ codons will be
represented by a hash (X).</li>
<li>In protein alignments, sequence following the termination codon,
will not be marked and will appear blank.</li>
<li>These conventions are used for both nucleotide and protein
alignments.</li>
</ul>
</blockquote>
<p>Learn more about the alignments at:</p>
<ul>
<li><a href="https://www.ebi.ac.uk/ipd/imgt/hla/alignment/help/" class="external-link uri">https://www.ebi.ac.uk/ipd/imgt/hla/alignment/help/</a></li>
</ul>
<p><code>a$aminos</code> has a matrix of amino acids with one column for
each position:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">a</span><span class="op">$</span><span class="va">aminos</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span>
<span><span class="co">#&gt;                  Pn29 Pn28 Pn27 Pn26 Pn25 Pn24 Pn23 Pn22 Pn21 Pn20</span></span>
<span><span class="co">#&gt; DRB1*01:01:01:01    M    V    C    L    K    L    P    G    G    S</span></span>
<span><span class="co">#&gt; DRB1*01:01:01:02    M    V    C    L    K    L    P    G    G    S</span></span>
<span><span class="co">#&gt; DRB1*01:01:01:03    M    V    C    L    K    L    P    G    G    S</span></span>
<span><span class="co">#&gt; DRB1*01:01:01:04    M    V    C    L    K    L    P    G    G    S</span></span>
<span><span class="co">#&gt; DRB1*01:01:01:05    M    V    C    L    K    L    P    G    G    S</span></span></code></pre></div>
<p><code>a$onehot</code> has a one-hot encoded matrix with one column
for each amino acid at each position:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">a</span><span class="op">$</span><span class="va">onehot</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">8</span><span class="op">]</span></span>
<span><span class="co">#&gt;                  Pn29_M Pn28_L Pn28_V Pn27_C Pn26_L Pn25_K Pn25_R Pn24_F</span></span>
<span><span class="co">#&gt; DRB1*01:01:01:01      1      0      1      1      1      1      0      0</span></span>
<span><span class="co">#&gt; DRB1*01:01:01:02      1      0      1      1      1      1      0      0</span></span>
<span><span class="co">#&gt; DRB1*01:01:01:03      1      0      1      1      1      1      0      0</span></span>
<span><span class="co">#&gt; DRB1*01:01:01:04      1      0      1      1      1      1      0      0</span></span>
<span><span class="co">#&gt; DRB1*01:01:01:05      1      0      1      1      1      1      0      0</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="convert-genotypes-to-a-dosage-matrix">Convert genotypes to a dosage matrix<a class="anchor" aria-label="anchor" href="#convert-genotypes-to-a-dosage-matrix"></a>
</h2>
<p>Suppose we have some individuals with the following genotypes:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genotypes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="st">"DRB1*12:02:02:03,DRB1*12:02:02:03,DRB1*14:54:02"</span>,</span>
<span>  <span class="st">"DRB1*04:174,DRB1*15:152"</span>,</span>
<span>  <span class="st">"DRB1*04:56:02,DRB1*15:01:48"</span>,</span>
<span>  <span class="st">"DRB1*14:172,DRB1*04:160"</span>,</span>
<span>  <span class="st">"DRB1*04:359,DRB1*04:284:02"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>If we want to run an association test on the amino acid positions,
then we need to convert the genotype names to a matrix of allele dosages
(e.g., 0, 1, 2).</p>
<p>We can use <code><a href="../reference/amino_dosage.html">amino_dosage()</a></code> to convert each individual’s
genotypes to amino acid dosages:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dosage</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/amino_dosage.html">amino_dosage</a></span><span class="op">(</span><span class="va">genotypes</span>, <span class="va">a</span><span class="op">$</span><span class="va">onehot</span><span class="op">)</span></span>
<span><span class="va">dosage</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span></span>
<span><span class="co">#&gt;                                                 Pn29_M P6_R P9_E P9_W P10_Q</span></span>
<span><span class="co">#&gt; DRB1*12:02:02:03,DRB1*12:02:02:03,DRB1*14:54:02      2    3    3    0     0</span></span>
<span><span class="co">#&gt; DRB1*04:174,DRB1*15:152                              0    2    1    1     2</span></span>
<span><span class="co">#&gt; DRB1*04:56:02,DRB1*15:01:48                          0    2    1    1     2</span></span>
<span><span class="co">#&gt; DRB1*14:172,DRB1*04:160                              0    2    2    0     1</span></span>
<span><span class="co">#&gt; DRB1*04:359,DRB1*04:284:02                           0    2    2    0     2</span></span>
<span><span class="co">#&gt;                                                 P10_Y</span></span>
<span><span class="co">#&gt; DRB1*12:02:02:03,DRB1*12:02:02:03,DRB1*14:54:02     3</span></span>
<span><span class="co">#&gt; DRB1*04:174,DRB1*15:152                             0</span></span>
<span><span class="co">#&gt; DRB1*04:56:02,DRB1*15:01:48                         0</span></span>
<span><span class="co">#&gt; DRB1*14:172,DRB1*04:160                             1</span></span>
<span><span class="co">#&gt; DRB1*04:359,DRB1*04:284:02                          0</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">dosage</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1]  5 35</span></span></code></pre></div>
<p><strong>Note:</strong></p>
<ul>
<li><p>The <code>dosage</code> matrix has one row for each individual
and one column for each amino acid at each position. By default,
<code><a href="../reference/amino_dosage.html">amino_dosage()</a></code> will discard the columns where all
individuals are identical.</p></li>
<li><p>The first individual has <code>dosage=3</code> for
<code>P6_R</code> (position 6 Arg). That’s because we assigned this
individual 3 alleles in our input.</p></li>
</ul>
<p>Please be careful to check that your data looks the way you
expect!</p>
</div>
<div class="section level2">
<h2 id="logistic-regression-association-for-amino-acid-positions">Logistic regression association for amino acid positions<a class="anchor" aria-label="anchor" href="#logistic-regression-association-for-amino-acid-positions"></a>
</h2>
<p>Let’s simulate a dataset with cases and controls to demonstrate one
approach for testing which amino acid positions might be associated with
cases.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">n</span> <span class="op">&lt;-</span> <span class="fl">100</span></span>
<span><span class="va">d</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  geno <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">a</span><span class="op">$</span><span class="va">onehot</span><span class="op">)</span>, <span class="va">n</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">a</span><span class="op">$</span><span class="va">onehot</span><span class="op">)</span>, <span class="va">n</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>    sep <span class="op">=</span> <span class="st">","</span></span>
<span>  <span class="op">)</span>,</span>
<span>  age <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">21</span><span class="op">:</span><span class="fl">100</span>, <span class="va">n</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>  case <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">0</span><span class="op">:</span><span class="fl">1</span>, <span class="va">n</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">d</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">d</span>, <span class="fu"><a href="../reference/amino_dosage.html">amino_dosage</a></span><span class="op">(</span><span class="va">d</span><span class="op">$</span><span class="va">geno</span>, <span class="va">a</span><span class="op">$</span><span class="va">onehot</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">d</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span></span>
<span><span class="co">#&gt;                                                  geno age case Pn29_M Pn25_K</span></span>
<span><span class="co">#&gt; DRB1*04:267N,DRB1*01:02:12 DRB1*04:267N,DRB1*01:02:12  73    1      0      0</span></span>
<span><span class="co">#&gt; DRB1*04:16,DRB1*15:02:06     DRB1*04:16,DRB1*15:02:06  72    0      1      1</span></span>
<span><span class="co">#&gt; DRB1*14:50:02,DRB1*04:288   DRB1*14:50:02,DRB1*04:288  22    1      0      0</span></span>
<span><span class="co">#&gt; DRB1*03:96,DRB1*03:106         DRB1*03:96,DRB1*03:106  34    0      0      0</span></span>
<span><span class="co">#&gt; DRB1*03:72,DRB1*13:51           DRB1*03:72,DRB1*13:51  76    0      0      0</span></span>
<span><span class="co">#&gt;                            Pn25_R</span></span>
<span><span class="co">#&gt; DRB1*04:267N,DRB1*01:02:12      0</span></span>
<span><span class="co">#&gt; DRB1*04:16,DRB1*15:02:06        0</span></span>
<span><span class="co">#&gt; DRB1*14:50:02,DRB1*04:288       0</span></span>
<span><span class="co">#&gt; DRB1*03:96,DRB1*03:106          0</span></span>
<span><span class="co">#&gt; DRB1*03:72,DRB1*13:51           0</span></span></code></pre></div>
<p>Our simulated dataset has 100 individuals, 52 cases and 48 controls.
We also have one column for each amino acid position that we might want
to test for association with the <code>case</code> variable.</p>
<p>One possible approach for association testing is to use
<code><a href="https://rdrr.io/r/stats/glm.html" class="external-link">glm()</a></code> to fit a logistic regression model for each amino
acid position. This could reveal if any amino acid position might be
associated with the <code>case</code> variable in our simulated
dataset.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># select the amino acid positions that have at least 3 people with dosage &gt; 0</span></span>
<span><span class="va">my_as</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">d</span><span class="op">[</span>,<span class="fl">4</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">d</span><span class="op">)</span><span class="op">]</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># run the association tests</span></span>
<span><span class="va">my_glm</span> <span class="op">&lt;-</span> <span class="fu">rbindlist</span><span class="op">(</span><span class="fu">pblapply</span><span class="op">(</span><span class="va">my_as</span>, <span class="kw">function</span><span class="op">(</span><span class="va">my_a</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">f</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"case ~ %s"</span>, <span class="va">my_a</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/glm.html" class="external-link">glm</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/formula.html" class="external-link">as.formula</a></span><span class="op">(</span><span class="va">f</span><span class="op">)</span>, data <span class="op">=</span> <span class="va">d</span>, family <span class="op">=</span> <span class="st">"binomial"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu">parameters</span><span class="op">(</span>exponentiate <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># look at the top hits</span></span>
<span><span class="va">my_glm</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">arrange</span><span class="op">(</span><span class="va">p</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="va">Parameter</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"(Intercept)"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="va">head</span></span>
<span><span class="co">#&gt;    Parameter Coefficient        SE   CI    CI_low    CI_high         z df_error</span></span>
<span><span class="co">#&gt; 1:    P101_E   4.7492150 2.8141197 0.95 1.6324022 17.5185505  2.629305      Inf</span></span>
<span><span class="co">#&gt; 2:     P34_H   3.0945832 1.3884885 0.95 1.3489859  7.9635257  2.517706      Inf</span></span>
<span><span class="co">#&gt; 3:     P13_H   2.7384213 1.1831088 0.95 1.2237664  6.7577338  2.331683      Inf</span></span>
<span><span class="co">#&gt; 4:     P34_N   0.3651739 0.1577699 0.95 0.1479786  0.8171494 -2.331683      Inf</span></span>
<span><span class="co">#&gt; 5:     P11_V   2.3962994 0.9557534 0.95 1.1336182  5.4854153  2.191138      Inf</span></span>
<span><span class="co">#&gt; 6:     P11_P   0.3636364 0.1786921 0.95 0.1329654  0.9307394 -2.058596      Inf</span></span>
<span><span class="co">#&gt;              p</span></span>
<span><span class="co">#&gt; 1: 0.008555961</span></span>
<span><span class="co">#&gt; 2: 0.011812185</span></span>
<span><span class="co">#&gt; 3: 0.019717351</span></span>
<span><span class="co">#&gt; 4: 0.019717351</span></span>
<span><span class="co">#&gt; 5: 0.028441812</span></span>
<span><span class="co">#&gt; 6: 0.039532948</span></span></code></pre></div>
<p>The volcano below shows the Odds Ratio and P-value for each amino
acid position. The top hits with P &lt; 0.05 are labeled.</p>
<p><img src="examples_files/figure-html/glm-volcano-1.png" width="576"></p>
<p>In this simulation, the <code>case</code> variable is associated with
P101_E (P = 0.0086, OR = 4.7, 95% CI 1.6 to 18).</p>
</div>
<div class="section level2">
<h2 id="umap-embedding-of-3516-hla-drb1-alleles">UMAP embedding of 3,516 HLA-DRB1 alleles<a class="anchor" aria-label="anchor" href="#umap-embedding-of-3516-hla-drb1-alleles"></a>
</h2>
<p>There are many things we might do with a one-hot encoding of HLA-DRB1
alleles.</p>
<p>For example, here is a UMAP embedding of 3,516 HLA-DRB1 alleles
encoded as a one-hot amino acid matrix with 1482 columns, one for each
amino acid at each position.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">uamp</span><span class="op">(</span><span class="va">a</span><span class="op">$</span><span class="va">onehot</span>, n_epochs <span class="op">=</span> <span class="fl">200</span>, min_dist <span class="op">=</span> <span class="fl">1</span>, spread <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="examples_files/figure-html/umap1-1.png" width="864"></p>
<p>We can highlight which alleles have amino acid H at position 13:</p>
<p><img src="examples_files/figure-html/umap-P13H-1.png" width="864"></p>
<p>Or we can represent each amino acid at position 13 with a different
color:</p>
<p><img src="examples_files/figure-html/umap-P13-1.png" width="864"></p>
</div>
<div class="section level2">
<h2 id="download-and-unpack-all-data-from-the-latest-imgthla-release">Download and unpack all data from the latest IMGTHLA release<a class="anchor" aria-label="anchor" href="#download-and-unpack-all-data-from-the-latest-imgthla-release"></a>
</h2>
<p>If you only want to use <code><a href="../reference/hla_alignments.html">hla_alignments()</a></code>, then you don’t
need <code><a href="../reference/install_hla.html">install_hla()</a></code> because data files are automatically
downloaded automatically as needed.</p>
<p>But some users might need access to additional files that are only
present in the full data release.</p>
<p>Run <code><a href="../reference/install_hla.html">install_hla()</a></code> to download and unpack the <a href="https://github.com/ANHIG/IMGTHLA/releases" class="external-link">latest IMGTHLA
release</a>. The destination folder for the downloaded data files is
<code>getOption("hlabud_dir")</code> (automatically tailored to your
operating system thanks to the <a href="https://github.com/r-lib/rappdirs" class="external-link">rappdirs</a> package).</p>
<p>Here are a few examples of how to download releases or get a list of
release names.</p>
<p>Download the latest release (default) or a specific release:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Download all of the data (120MB) for the latest IMGTHLA release</span></span>
<span><span class="fu"><a href="../reference/install_hla.html">install_hla</a></span><span class="op">(</span>release <span class="op">=</span> <span class="st">"latest"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Download a specific release</span></span>
<span><span class="fu"><a href="../reference/install_hla.html">install_hla</a></span><span class="op">(</span>release <span class="op">=</span> <span class="st">"3.51.0"</span><span class="op">)</span></span></code></pre></div>
<p>Optionally, get or set the directory hlabud uses to store the
data:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"hlabud_dir"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "/home/username/.local/share/hlabud"</span></span>
<span></span>
<span><span class="co"># Manually override the directory for hlabud to use</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>hlabud_dir <span class="op">=</span> <span class="st">"/path/to/my/dir"</span><span class="op">)</span></span></code></pre></div>
<p>Check which release hlabud is using, or choose a release:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"hlabud_release"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "3.51.0"</span></span>
<span></span>
<span><span class="co"># Use a specific release</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>hlabud_release <span class="op">=</span> <span class="st">"3.51.0"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># List all releases</span></span>
<span><span class="fu"><a href="../reference/hla_releases.html">hla_releases</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "3.51.0"   "3.50.0"   "3.49.0"   "3.48.0"   "3.47.0"   "3.46.0"   "3.45.1"   "3.45.01"</span></span>
<span><span class="co">#&gt;  [9] "3.45.0.1" "3.45.0"   "3.44.1"   "3.44.0"   "3.43.0"   "3.42.0"   "3.41.2"   "3.41.0"</span></span>
<span><span class="co">#&gt; [17] "3.40.0"   "3.39.0"   "3.38.0"   "3.37.0"   "3.36.0"   "3.35.0"   "3.34.0"   "3.33.0"</span></span>
<span><span class="co">#&gt; [25] "3.32.0"   "3.31.0"   "3.30.0"   "3.29.0"   "3.28.0"   "3.27.0"</span></span></code></pre></div>
<p>After installing a few releases, the hlabud folder might look like
this:</p>
<pre><code>❯ ls -lah "/home/user/.local/share/hlabud"
total 207M
drwxrwxr-x  3 user user      32 Apr  5 01:19 3.30.0
drwxrwxr-x 11 user user    4.0K Apr  7 19:31 3.40.0
drwxrwxr-x 12 user user    4.0K Apr  5 00:27 3.51.0
-rw-rw-r--  1 user user     15K Apr  7 19:23 tags.json
-rw-rw-r--  1 user user     79M Apr  7 19:28 v3.40.0-alpha.tar.gz
-rw-rw-r--  1 user user    129M Apr  4 20:07 v3.51.0-alpha.tar.gz</code></pre>
</div>
<div class="section level2">
<h2 id="count-the-number-of-alleles-in-each-imgthla-release">Count the number of alleles in each IMGTHLA release<a class="anchor" aria-label="anchor" href="#count-the-number-of-alleles-in-each-imgthla-release"></a>
</h2>
<p>We can get a list of the release names:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">releases</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hla_releases.html">hla_releases</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">releases</span></span>
<span><span class="co">#&gt;  [1] "3.52.0"   "3.51.0"   "3.50.0"   "3.49.0"   "3.48.0"   "3.47.0"  </span></span>
<span><span class="co">#&gt;  [7] "3.46.0"   "3.45.1"   "3.45.01"  "3.45.0.1" "3.45.0"   "3.44.1"  </span></span>
<span><span class="co">#&gt; [13] "3.44.0"   "3.43.0"   "3.42.0"   "3.41.2"   "3.41.0"   "3.40.0"  </span></span>
<span><span class="co">#&gt; [19] "3.39.0"   "3.38.0"   "3.37.0"   "3.36.0"   "3.35.0"   "3.34.0"  </span></span>
<span><span class="co">#&gt; [25] "3.33.0"   "3.32.0"   "3.31.0"   "3.30.0"   "3.29.0"   "3.28.0"</span></span></code></pre></div>
<p>Then we can get the allele names for each release:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">my_alleles</span> <span class="op">&lt;-</span> <span class="fu">rbindlist</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">releases</span>, <span class="kw">function</span><span class="op">(</span><span class="va">release</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">retval</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hla_alleles.html">hla_alleles</a></span><span class="op">(</span>release <span class="op">=</span> <span class="va">release</span><span class="op">)</span></span>
<span>  <span class="va">retval</span><span class="op">$</span><span class="va">release</span> <span class="op">&lt;-</span> <span class="va">release</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">retval</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span>, fill <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in hla_alleles(release = release): unrecognized release name</span></span>
<span><span class="co">#&gt; 'Allelelist.3451.txt'</span></span>
<span><span class="co">#&gt; Warning in hla_alleles(release = release): unrecognized release name</span></span>
<span><span class="co">#&gt; 'Allelelist.34501.txt'</span></span>
<span></span>
<span><span class="co">#&gt; Warning in hla_alleles(release = release): unrecognized release name</span></span>
<span><span class="co">#&gt; 'Allelelist.34501.txt'</span></span>
<span><span class="co">#&gt; Warning in hla_alleles(release = release): unrecognized release name</span></span>
<span><span class="co">#&gt; 'Allelelist.3441.txt'</span></span>
<span><span class="co">#&gt; Warning in hla_alleles(release = release): unrecognized release name</span></span>
<span><span class="co">#&gt; 'Allelelist.3412.txt'</span></span></code></pre></div>
<p>Next, count how many alleles we have in each release:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">d</span> <span class="op">&lt;-</span> <span class="va">my_alleles</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu">count</span><span class="op">(</span><span class="va">release</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">n</span> <span class="op">&gt;</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">d</span></span>
<span><span class="co">#&gt;     release     n</span></span>
<span><span class="co">#&gt;  1:  3.28.0 16933</span></span>
<span><span class="co">#&gt;  2:  3.29.0 17344</span></span>
<span><span class="co">#&gt;  3:  3.30.0 17509</span></span>
<span><span class="co">#&gt;  4:  3.31.0 17874</span></span>
<span><span class="co">#&gt;  5:  3.32.0 18363</span></span>
<span><span class="co">#&gt;  6:  3.33.0 18955</span></span>
<span><span class="co">#&gt;  7:  3.34.0 20272</span></span>
<span><span class="co">#&gt;  8:  3.35.0 21683</span></span>
<span><span class="co">#&gt;  9:  3.36.0 22548</span></span>
<span><span class="co">#&gt; 10:  3.37.0 24093</span></span>
<span><span class="co">#&gt; 11:  3.38.0 25958</span></span>
<span><span class="co">#&gt; 12:  3.39.0 26512</span></span>
<span><span class="co">#&gt; 13:  3.40.0 27273</span></span>
<span><span class="co">#&gt; 14:  3.41.0 27980</span></span>
<span><span class="co">#&gt; 15:  3.42.0 28786</span></span>
<span><span class="co">#&gt; 16:  3.43.0 29417</span></span>
<span><span class="co">#&gt; 17:  3.44.0 30523</span></span>
<span><span class="co">#&gt; 18:  3.45.0 31552</span></span>
<span><span class="co">#&gt; 19:  3.46.0 32330</span></span>
<span><span class="co">#&gt; 20:  3.47.0 33552</span></span>
<span><span class="co">#&gt; 21:  3.48.0 34145</span></span>
<span><span class="co">#&gt; 22:  3.49.0 35077</span></span>
<span><span class="co">#&gt; 23:  3.50.0 36016</span></span>
<span><span class="co">#&gt; 24:  3.51.0 36625</span></span>
<span><span class="co">#&gt; 25:  3.52.0 37068</span></span>
<span><span class="co">#&gt;     release     n</span></span></code></pre></div>
<p>And plot the number of alleles as a line plot:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span><span class="va">d</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">release</span>, y <span class="op">=</span> <span class="va">n</span>, group <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_line</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_text</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>label <span class="op">=</span> <span class="va">release</span><span class="op">)</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span>x <span class="op">=</span> <span class="cn">NULL</span>, y <span class="op">=</span> <span class="st">"Number of alleles"</span>,</span>
<span>  title <span class="op">=</span> <span class="st">"Each release has more HLA alleles"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span></span>
<span>    axis.text.x <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>,</span>
<span>    axis.ticks.x <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="examples_files/figure-html/unnamed-chunk-20-1.png" width="700"></p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">d2</span> <span class="op">&lt;-</span> <span class="va">my_alleles</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu">mutate</span><span class="op">(</span>gene <span class="op">=</span> <span class="fu">str_split_fixed</span><span class="op">(</span><span class="va">Allele</span>, <span class="st">"\\*"</span>, <span class="fl">2</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu">count</span><span class="op">(</span><span class="va">release</span>, <span class="va">gene</span><span class="op">)</span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">release</span>, y <span class="op">=</span> <span class="va">n</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_line</span><span class="op">(</span></span>
<span>    data <span class="op">=</span> <span class="va">d2</span>,</span>
<span>    <span class="fu">aes</span><span class="op">(</span>group <span class="op">=</span> <span class="va">gene</span>, color <span class="op">=</span> <span class="va">gene</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">scale_color_discrete</span><span class="op">(</span>guide <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_text</span><span class="op">(</span></span>
<span>    data <span class="op">=</span> <span class="va">d2</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">release</span> <span class="op">==</span> <span class="st">"3.52.0"</span><span class="op">)</span>,</span>
<span>    mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>label <span class="op">=</span> <span class="va">gene</span><span class="op">)</span>,</span>
<span>    hjust <span class="op">=</span> <span class="fl">0</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span>x <span class="op">=</span> <span class="cn">NULL</span>, y <span class="op">=</span> <span class="st">"Number of alleles"</span>,</span>
<span>  title <span class="op">=</span> <span class="st">"Number of alleles per release and gene"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">scale_x_discrete</span><span class="op">(</span>expand <span class="op">=</span> <span class="fu">expansion</span><span class="op">(</span>mult <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.01</span>, <span class="fl">0.1</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">scale_y_log10</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span></span>
<span>    panel.grid.major.y <span class="op">=</span> <span class="fu">element_line</span><span class="op">(</span><span class="op">)</span>, </span>
<span>    axis.text.x <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>,</span>
<span>    axis.ticks.x <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="examples_files/figure-html/unnamed-chunk-21-1.png" width="700"></p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Kamil Slowikowski.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
