<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>hlabud usage examples • hlabud</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="hlabud usage examples">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">hlabud</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/examples.html">hlabud usage examples</a></li>
    <li><a class="dropdown-item" href="../articles/numbering.html">Numbering amino acid positions</a></li>
    <li><a class="dropdown-item" href="../articles/visualize-hla-structure.html">Visualize HLA protein structures</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/slowkow/hlabud"><span class="fa fa-github"></span> GitHub</a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>hlabud usage examples</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/slowkow/hlabud/blob/main/vignettes/articles/examples.Rmd" class="external-link"><code>vignettes/articles/examples.Rmd</code></a></small>
      <div class="d-none name"><code>examples.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Kamil Slowikowski</p>
<p>2025-02-06</p>
<p><a href="https://github.com/slowkow/hlabud" class="external-link">hlabud</a> is an R
package that provides functions to facilitate download and analysis of
human leukocyte antigen (HLA) genotype sequence alignments from <a href="https://github.com/ANHIG/IMGTHLA" class="external-link">IMGTHLA</a> in R.</p>
<p>Let’s consider a question that we might want to answer about HLA
genotypes.</p>
<p>What amino acid positions are different between two genotypes?</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/slowkow/hlabud" class="external-link">hlabud</a></span><span class="op">)</span></span>
<span><span class="va">a</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hla_alignments.html">hla_alignments</a></span><span class="op">(</span><span class="st">"DRB1"</span><span class="op">)</span></span>
<span><span class="va">a</span><span class="op">$</span><span class="va">release</span></span>
<span><span class="co">#&gt; [1] "3.59.0"</span></span>
<span><span class="fu"><a href="../reference/dosage.html">dosage</a></span><span class="op">(</span><span class="va">a</span><span class="op">$</span><span class="va">onehot</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"DRB1*03:01:05"</span>, <span class="st">"DRB1*03:02:03"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;               F26 Y26 D28 E28 F47 Y47 G86 V86</span></span>
<span><span class="co">#&gt; DRB1*03:01:05   0   1   1   0   1   0   0   1</span></span>
<span><span class="co">#&gt; DRB1*03:02:03   1   0   0   1   0   1   1   0</span></span></code></pre></div>
<p>What nucleotides are different?</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">n</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hla_alignments.html">hla_alignments</a></span><span class="op">(</span><span class="st">"DRB1"</span>, type <span class="op">=</span> <span class="st">"nuc"</span><span class="op">)</span></span>
<span><span class="va">n</span><span class="op">$</span><span class="va">release</span></span>
<span><span class="co">#&gt; [1] "3.59.0"</span></span>
<span><span class="fu"><a href="../reference/dosage.html">dosage</a></span><span class="op">(</span><span class="va">n</span><span class="op">$</span><span class="va">onehot</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"DRB1*03:01:05"</span>, <span class="st">"DRB1*03:02:03"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;               A164 T164 C171 G171 A227 T227 A240 G240 G344 T344 G345 T345 A357</span></span>
<span><span class="co">#&gt; DRB1*03:01:05    1    0    1    0    0    1    1    0    0    1    1    0    1</span></span>
<span><span class="co">#&gt; DRB1*03:02:03    0    1    0    1    1    0    0    1    1    0    0    1    0</span></span>
<span><span class="co">#&gt;               G357</span></span>
<span><span class="co">#&gt; DRB1*03:01:05    0</span></span>
<span><span class="co">#&gt; DRB1*03:02:03    1</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>The quickest way to get hlabud is to install from GitHub:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install.packages("devtools")</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"slowkow/hlabud"</span><span class="op">)</span></span></code></pre></div>
<p>Below, I included a few usage examples. I hope they inspire you to
share your own HLA analyses.</p>
<p>The source code for this page is available <a href="https://github.com/slowkow/hlabud/tree/main/vignettes/examples.Rmd" class="external-link">here</a>.</p>
<p>Thank you for <a href="https://github.com/slowkow/hlabud/issues" class="external-link">reporting issues</a>
with hlabud.</p>
</div>
<div class="section level2">
<h2 id="get-a-one-hot-encoded-matrix-for-all-hla-drb1-alleles">Get a one-hot encoded matrix for all HLA-DRB1 alleles<a class="anchor" aria-label="anchor" href="#get-a-one-hot-encoded-matrix-for-all-hla-drb1-alleles"></a>
</h2>
<p>We can use <code>hla_alignments("DRB1")</code> to load the
<code>DRB1_prot.txt</code> file from the latest <a href="https://github.com/ANHIG/IMGTHLA/releases" class="external-link">IMGTHLA</a>
release:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/slowkow/hlabud" class="external-link">hlabud</a></span><span class="op">)</span></span>
<span><span class="va">a</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hla_alignments.html">hla_alignments</a></span><span class="op">(</span>gene <span class="op">=</span> <span class="st">"DRB1"</span>, verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Reading /home/runner/.local/share/hlabud/3.59.0/alignments/DRB1_prot.txt</span></span></code></pre></div>
<p>The <code>a</code> object is a list with three items:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">a</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 7</span></span>
<span><span class="co">#&gt;  $ sequences: Named chr [1:3786] "MVCLKLPGGSCMTALTVTLMVLSSPLALAGDTRPRFLWQLKFECHFFNGTERVR.LLERCIYNQEE.SVRFDSDVGEYRAVTELGRPDAEYWNSQKDLLEQRRAAVDTYCR"| __truncated__ "------------------------------------------------------.-----------.--------------------------------------------"| __truncated__ "------------------------------------------------------.-----------.--------------------------------------------"| __truncated__ "------------------------------------------------------.-----------.--------------------------------------------"| __truncated__ ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "names")= chr [1:3786] "DRB1*01:01:01:01" "DRB1*01:01:01:02" "DRB1*01:01:01:03" "DRB1*01:01:01:04" ...</span></span>
<span><span class="co">#&gt;  $ alleles  : chr [1:3786, 1:288] "M" "M" "M" "M" ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">#&gt;   .. ..$ : chr [1:3786] "DRB1*01:01:01:01" "DRB1*01:01:01:02" "DRB1*01:01:01:03" "DRB1*01:01:01:04" ...</span></span>
<span><span class="co">#&gt;   .. ..$ : chr [1:288] "n29" "n28" "n27" "n26" ...</span></span>
<span><span class="co">#&gt;  $ onehot   : num [1:3786, 1:1725] 0 0 0 0 0 0 0 0 0 0 ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">#&gt;   .. ..$ : chr [1:3786] "DRB1*01:01:01:01" "DRB1*01:01:01:02" "DRB1*01:01:01:03" "DRB1*01:01:01:04" ...</span></span>
<span><span class="co">#&gt;   .. ..$ : chr [1:1725] "n29unk" "Mn29" "n28unk" "Ln28" ...</span></span>
<span><span class="co">#&gt;  $ gene     : chr "DRB1"</span></span>
<span><span class="co">#&gt;  $ type     : chr "prot"</span></span>
<span><span class="co">#&gt;  $ release  : chr "3.59.0"</span></span>
<span><span class="co">#&gt;  $ file     : chr "/home/runner/.local/share/hlabud/3.59.0/alignments/DRB1_prot.txt"</span></span></code></pre></div>
<p><code>a$sequences</code> has amino acid sequence alignments in a
named character vector:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/substr.html" class="external-link">substr</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">a</span><span class="op">$</span><span class="va">sequences</span>, <span class="fl">6</span><span class="op">)</span>, <span class="fl">1</span>, <span class="fl">50</span><span class="op">)</span></span>
<span><span class="co">#&gt;                                     DRB1*01:01:01:01 </span></span>
<span><span class="co">#&gt; "MVCLKLPGGSCMTALTVTLMVLSSPLALAGDTRPRFLWQLKFECHFFNGT" </span></span>
<span><span class="co">#&gt;                                     DRB1*01:01:01:02 </span></span>
<span><span class="co">#&gt; "--------------------------------------------------" </span></span>
<span><span class="co">#&gt;                                     DRB1*01:01:01:03 </span></span>
<span><span class="co">#&gt; "--------------------------------------------------" </span></span>
<span><span class="co">#&gt;                                     DRB1*01:01:01:04 </span></span>
<span><span class="co">#&gt; "--------------------------------------------------" </span></span>
<span><span class="co">#&gt;                                     DRB1*01:01:01:05 </span></span>
<span><span class="co">#&gt; "--------------------------------------------------" </span></span>
<span><span class="co">#&gt;                                     DRB1*01:01:01:06 </span></span>
<span><span class="co">#&gt; "--------------------------------------------------"</span></span></code></pre></div>
<p>Here are the conventions used for alignments (copied from the EBI <a href="https://www.ebi.ac.uk/ipd/imgt/hla/alignment/help/" class="external-link">help
page</a>):</p>
<blockquote>
<ul>
<li>The entry for each allele is displayed in respect to the reference
sequences.</li>
<li>Where identity to the reference sequence is present the base will be
displayed as a hyphen (-).</li>
<li>Non-identity to the reference sequence is shown by displaying the
appropriate base at that position.</li>
<li>Where an insertion or deletion has occurred this will be represented
by a period (.).</li>
<li>If the sequence is unknown at any point in the alignment, this will
be represented by an asterisk (*).</li>
<li>In protein alignments for null alleles, the ‘Stop’ codons will be
represented by a hash (X).</li>
<li>In protein alignments, sequence following the termination codon,
will not be marked and will appear blank.</li>
<li>These conventions are used for both nucleotide and protein
alignments.</li>
</ul>
</blockquote>
<p><code>a$alleles</code> has a matrix of amino acids with one column
for each position:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">a</span><span class="op">$</span><span class="va">alleles</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">40</span><span class="op">]</span></span>
<span><span class="co">#&gt;                  n29 n28 n27 n26 n25 n24 n23 n22 n21 n20 n19 n18 n17 n16 n15</span></span>
<span><span class="co">#&gt; DRB1*01:01:01:01 "M" "V" "C" "L" "K" "L" "P" "G" "G" "S" "C" "M" "T" "A" "L"</span></span>
<span><span class="co">#&gt; DRB1*01:01:01:02 "M" "V" "C" "L" "K" "L" "P" "G" "G" "S" "C" "M" "T" "A" "L"</span></span>
<span><span class="co">#&gt; DRB1*01:01:01:03 "M" "V" "C" "L" "K" "L" "P" "G" "G" "S" "C" "M" "T" "A" "L"</span></span>
<span><span class="co">#&gt; DRB1*01:01:01:04 "M" "V" "C" "L" "K" "L" "P" "G" "G" "S" "C" "M" "T" "A" "L"</span></span>
<span><span class="co">#&gt; DRB1*01:01:01:05 "M" "V" "C" "L" "K" "L" "P" "G" "G" "S" "C" "M" "T" "A" "L"</span></span>
<span><span class="co">#&gt;                  n14 n13 n12 n11 n10 n9  n8  n7  n6  n5  n4  n3  n2  n1  1  </span></span>
<span><span class="co">#&gt; DRB1*01:01:01:01 "T" "V" "T" "L" "M" "V" "L" "S" "S" "P" "L" "A" "L" "A" "G"</span></span>
<span><span class="co">#&gt; DRB1*01:01:01:02 "T" "V" "T" "L" "M" "V" "L" "S" "S" "P" "L" "A" "L" "A" "G"</span></span>
<span><span class="co">#&gt; DRB1*01:01:01:03 "T" "V" "T" "L" "M" "V" "L" "S" "S" "P" "L" "A" "L" "A" "G"</span></span>
<span><span class="co">#&gt; DRB1*01:01:01:04 "T" "V" "T" "L" "M" "V" "L" "S" "S" "P" "L" "A" "L" "A" "G"</span></span>
<span><span class="co">#&gt; DRB1*01:01:01:05 "T" "V" "T" "L" "M" "V" "L" "S" "S" "P" "L" "A" "L" "A" "G"</span></span>
<span><span class="co">#&gt;                  2   3   4   5   6   7   8   9   10  11 </span></span>
<span><span class="co">#&gt; DRB1*01:01:01:01 "D" "T" "R" "P" "R" "F" "L" "W" "Q" "L"</span></span>
<span><span class="co">#&gt; DRB1*01:01:01:02 "D" "T" "R" "P" "R" "F" "L" "W" "Q" "L"</span></span>
<span><span class="co">#&gt; DRB1*01:01:01:03 "D" "T" "R" "P" "R" "F" "L" "W" "Q" "L"</span></span>
<span><span class="co">#&gt; DRB1*01:01:01:04 "D" "T" "R" "P" "R" "F" "L" "W" "Q" "L"</span></span>
<span><span class="co">#&gt; DRB1*01:01:01:05 "D" "T" "R" "P" "R" "F" "L" "W" "Q" "L"</span></span></code></pre></div>
<p><code>a$onehot</code> has a one-hot encoded matrix with one column
for each amino acid at each position:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">a</span><span class="op">$</span><span class="va">onehot</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">25</span><span class="op">]</span></span>
<span><span class="co">#&gt;                  n29unk Mn29 n28unk Ln28 Vn28 n27unk Cn27 n26unk Ln26 n25unk</span></span>
<span><span class="co">#&gt; DRB1*01:01:01:01      0    1      0    0    1      0    1      0    1      0</span></span>
<span><span class="co">#&gt; DRB1*01:01:01:02      0    1      0    0    1      0    1      0    1      0</span></span>
<span><span class="co">#&gt; DRB1*01:01:01:03      0    1      0    0    1      0    1      0    1      0</span></span>
<span><span class="co">#&gt; DRB1*01:01:01:04      0    1      0    0    1      0    1      0    1      0</span></span>
<span><span class="co">#&gt; DRB1*01:01:01:05      0    1      0    0    1      0    1      0    1      0</span></span>
<span><span class="co">#&gt;                  Kn25 Rn25 Sn25 n24unk Fn24 Ln24 n23unk Pn23 n22unk Gn22 n21unk</span></span>
<span><span class="co">#&gt; DRB1*01:01:01:01    1    0    0      0    0    1      0    1      0    1      0</span></span>
<span><span class="co">#&gt; DRB1*01:01:01:02    1    0    0      0    0    1      0    1      0    1      0</span></span>
<span><span class="co">#&gt; DRB1*01:01:01:03    1    0    0      0    0    1      0    1      0    1      0</span></span>
<span><span class="co">#&gt; DRB1*01:01:01:04    1    0    0      0    0    1      0    1      0    1      0</span></span>
<span><span class="co">#&gt; DRB1*01:01:01:05    1    0    0      0    0    1      0    1      0    1      0</span></span>
<span><span class="co">#&gt;                  Cn21 Gn21 n20unk Sn20</span></span>
<span><span class="co">#&gt; DRB1*01:01:01:01    0    1      0    1</span></span>
<span><span class="co">#&gt; DRB1*01:01:01:02    0    1      0    1</span></span>
<span><span class="co">#&gt; DRB1*01:01:01:03    0    1      0    1</span></span>
<span><span class="co">#&gt; DRB1*01:01:01:04    0    1      0    1</span></span>
<span><span class="co">#&gt; DRB1*01:01:01:05    0    1      0    1</span></span></code></pre></div>
<p>What is a one-hot encoded matrix? Here is a simple example to
demonstrate the idea:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  V1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"A"</span>, <span class="st">"B"</span><span class="op">)</span>,</span>
<span>  V2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"B"</span>, <span class="st">"B"</span>, <span class="st">"B"</span><span class="op">)</span>,</span>
<span>  V3 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"C"</span>, <span class="st">"B"</span>, <span class="st">"B"</span><span class="op">)</span>,</span>
<span>  stringsAsFactors <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="va">dat</span></span>
<span><span class="co">#&gt;   V1 V2 V3</span></span>
<span><span class="co">#&gt; 1  A  B  C</span></span>
<span><span class="co">#&gt; 2  A  B  B</span></span>
<span><span class="co">#&gt; 3  B  B  B</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict</a></span><span class="op">(</span><span class="fu">onehot</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/onehot/man/onehot.html" class="external-link">onehot</a></span><span class="op">(</span><span class="va">dat</span><span class="op">)</span>, <span class="va">dat</span><span class="op">)</span></span>
<span><span class="co">#&gt;      V1=A V1=B V2=B V3=B V3=C</span></span>
<span><span class="co">#&gt; [1,]    1    0    1    0    1</span></span>
<span><span class="co">#&gt; [2,]    1    0    1    1    0</span></span>
<span><span class="co">#&gt; [3,]    0    1    1    1    0</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="convert-genotypes-to-a-dosage-matrix">Convert genotypes to a dosage matrix<a class="anchor" aria-label="anchor" href="#convert-genotypes-to-a-dosage-matrix"></a>
</h2>
<p>Suppose we have some individuals with the following genotypes:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genotypes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="st">"DRB1*12:02:02:03,DRB1*12:02:02:03"</span>,</span>
<span>  <span class="st">"DRB1*04:174,DRB1*15:152"</span>,</span>
<span>  <span class="st">"DRB1*04:56:02,DRB1*15:01:48"</span>,</span>
<span>  <span class="st">"DRB1*14:172,DRB1*04:160"</span>,</span>
<span>  <span class="st">"DRB1*04:359,DRB1*04:284:02"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>If we want to run an association test on the amino acid positions,
then we need to convert the genotype names to a matrix of allele dosages
(e.g., 0, 1, 2).</p>
<p>We can use <code><a href="../reference/dosage.html">dosage()</a></code> to convert each individual’s
genotypes to amino acid dosages:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dosage</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dosage.html">dosage</a></span><span class="op">(</span><span class="va">a</span><span class="op">$</span><span class="va">onehot</span>, <span class="va">genotypes</span><span class="op">)</span></span>
<span><span class="va">dosage</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">8</span><span class="op">]</span></span>
<span><span class="co">#&gt;                                   n29unk Mn29 n28unk Vn28 n27unk Cn27 n26unk</span></span>
<span><span class="co">#&gt; DRB1*12:02:02:03,DRB1*12:02:02:03      0    2      0    2      0    2      0</span></span>
<span><span class="co">#&gt; DRB1*04:174,DRB1*15:152                2    0      2    0      2    0      2</span></span>
<span><span class="co">#&gt; DRB1*04:56:02,DRB1*15:01:48            2    0      2    0      2    0      2</span></span>
<span><span class="co">#&gt; DRB1*14:172,DRB1*04:160                2    0      2    0      2    0      2</span></span>
<span><span class="co">#&gt; DRB1*04:359,DRB1*04:284:02             2    0      2    0      2    0      2</span></span>
<span><span class="co">#&gt;                                   Ln26</span></span>
<span><span class="co">#&gt; DRB1*12:02:02:03,DRB1*12:02:02:03    2</span></span>
<span><span class="co">#&gt; DRB1*04:174,DRB1*15:152              0</span></span>
<span><span class="co">#&gt; DRB1*04:56:02,DRB1*15:01:48          0</span></span>
<span><span class="co">#&gt; DRB1*14:172,DRB1*04:160              0</span></span>
<span><span class="co">#&gt; DRB1*04:359,DRB1*04:284:02           0</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">dosage</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1]   5 428</span></span></code></pre></div>
<p><strong>Note:</strong></p>
<ul>
<li><p>The <code>dosage</code> matrix has one row for each individual
and one column for each amino acid at each position. By default,
<code><a href="../reference/dosage.html">dosage()</a></code> will discard the columns where all individuals are
identical.</p></li>
<li><p>If input allele names are truncated to 4-digits or 2-digits (e.g.
<code>DRB1*03:01</code> or <code>DRB1*03</code>), then hlabud will pick
the first allele that matches the input allele
(e.g. <code>DRB1*03:01:01:01</code>). If you want a specific allele,
then you need to provide the full allele name in the input.</p></li>
</ul>
<p>Please be careful to check that your data looks the way you
expect!</p>
</div>
<div class="section level2">
<h2 id="logistic-regression-association-for-amino-acid-positions">Logistic regression association for amino acid positions<a class="anchor" aria-label="anchor" href="#logistic-regression-association-for-amino-acid-positions"></a>
</h2>
<p>Let’s simulate a dataset with cases and controls to demonstrate one
approach for testing which amino acid positions might be associated with
cases.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">n</span> <span class="op">&lt;-</span> <span class="fl">100</span></span>
<span><span class="va">d</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  geno <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">a</span><span class="op">$</span><span class="va">onehot</span><span class="op">)</span>, <span class="va">n</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">a</span><span class="op">$</span><span class="va">onehot</span><span class="op">)</span>, <span class="va">n</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>    sep <span class="op">=</span> <span class="st">","</span></span>
<span>  <span class="op">)</span>,</span>
<span>  age <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">21</span><span class="op">:</span><span class="fl">100</span>, <span class="va">n</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>  case <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">0</span><span class="op">:</span><span class="fl">1</span>, <span class="va">n</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">d</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">d</span>, <span class="fu"><a href="../reference/dosage.html">dosage</a></span><span class="op">(</span><span class="va">a</span><span class="op">$</span><span class="va">onehot</span>, <span class="va">d</span><span class="op">$</span><span class="va">geno</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">d</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span></span>
<span><span class="co">#&gt;                                                geno age case n29unk Mn29 n28unk</span></span>
<span><span class="co">#&gt; DRB1*04:224,DRB1*09:29       DRB1*04:224,DRB1*09:29  73    0      2    0      2</span></span>
<span><span class="co">#&gt; DRB1*04:05:23,DRB1*14:183 DRB1*04:05:23,DRB1*14:183  79    0      2    0      2</span></span>
<span><span class="co">#&gt; DRB1*13:263,DRB1*04:40       DRB1*13:263,DRB1*04:40  67    0      2    0      2</span></span>
<span><span class="co">#&gt; DRB1*03:75,DRB1*04:95:02   DRB1*03:75,DRB1*04:95:02  38    1      2    0      2</span></span>
<span><span class="co">#&gt; DRB1*03:53,DRB1*01:02:09   DRB1*03:53,DRB1*01:02:09  67    0      2    0      2</span></span></code></pre></div>
<p>Our simulated dataset has 100 individuals, 51 cases and 49 controls.
We also have one column for each amino acid position that we might want
to test for association with the <code>case</code> variable.</p>
<p>One possible approach for association testing is to use
<code><a href="https://rdrr.io/r/stats/glm.html" class="external-link">glm()</a></code> to fit a logistic regression model for each amino
acid position. This could reveal if any amino acid position might be
associated with the <code>case</code> variable in our simulated
dataset.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># prepare column names for use in formulas</span></span>
<span><span class="va">ix</span> <span class="op">&lt;-</span> <span class="fl">4</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">d</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">d</span><span class="op">)</span><span class="op">[</span><span class="va">ix</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"VAR%s"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">d</span><span class="op">)</span><span class="op">[</span><span class="va">ix</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># select the amino acid positions that have at least 3 people with dosage &gt; 0</span></span>
<span><span class="va">my_as</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">d</span><span class="op">[</span>,<span class="fl">4</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">d</span><span class="op">)</span><span class="op">]</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># run the association tests</span></span>
<span><span class="va">my_glm</span> <span class="op">&lt;-</span> <span class="fu">rbindlist</span><span class="op">(</span><span class="fu">pblapply</span><span class="op">(</span><span class="va">my_as</span>, <span class="kw">function</span><span class="op">(</span><span class="va">my_a</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">f</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"case ~ %s"</span>, <span class="va">my_a</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/glm.html" class="external-link">glm</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/formula.html" class="external-link">as.formula</a></span><span class="op">(</span><span class="va">f</span><span class="op">)</span>, data <span class="op">=</span> <span class="va">d</span>, family <span class="op">=</span> <span class="st">"binomial"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu">parameters</span><span class="op">(</span>exponentiate <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># look at the top hits</span></span>
<span><span class="va">my_glm</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">arrange</span><span class="op">(</span><span class="va">p</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="va">Parameter</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"(Intercept)"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="va">head</span></span>
<span><span class="co">#&gt;    Parameter Coefficient        SE    CI    CI_low   CI_high         z df_error</span></span>
<span><span class="co">#&gt;       &lt;char&gt;       &lt;num&gt;     &lt;num&gt; &lt;num&gt;     &lt;num&gt;     &lt;num&gt;     &lt;num&gt;    &lt;num&gt;</span></span>
<span><span class="co">#&gt; 1:   VARR233   0.3740889 0.1537387  0.95 0.1585149 0.8055764 -2.392548      Inf</span></span>
<span><span class="co">#&gt; 2:    VARQ10   2.0504158 0.6343398  0.95 1.1365450 3.8562617  2.320974      Inf</span></span>
<span><span class="co">#&gt; 3:    VART12   0.5050496 0.1574155  0.95 0.2673406 0.9158840 -2.191644      Inf</span></span>
<span><span class="co">#&gt; 4:    VARK12   1.9800034 0.6171339  0.95 1.0918414 3.7405461  2.191644      Inf</span></span>
<span><span class="co">#&gt; 5:    VARS11   0.5086017 0.1578058  0.95 0.2700494 0.9200341 -2.179011      Inf</span></span>
<span><span class="co">#&gt; 6:    VARY10   0.5239755 0.1638312  0.95 0.2770470 0.9529264 -2.067071      Inf</span></span>
<span><span class="co">#&gt;             p</span></span>
<span><span class="co">#&gt;         &lt;num&gt;</span></span>
<span><span class="co">#&gt; 1: 0.01673184</span></span>
<span><span class="co">#&gt; 2: 0.02028826</span></span>
<span><span class="co">#&gt; 3: 0.02840525</span></span>
<span><span class="co">#&gt; 4: 0.02840525</span></span>
<span><span class="co">#&gt; 5: 0.02933082</span></span>
<span><span class="co">#&gt; 6: 0.03872745</span></span></code></pre></div>
<p>The volcano below shows the Odds Ratio and P-value for each amino
acid position. The top hits with P &lt; 0.05 are labeled.</p>
<p><img src="examples_files/figure-html/glm-volcano-1.png" width="1800"></p>
<p>In this simulation, the <code>case</code> variable is associated with
R233 (P = 0.017, OR = 0.37, 95% CI 0.16 to 0.81).</p>
</div>
<div class="section level2">
<h2 id="umap-embedding-of-hla-drb1-alleles">UMAP embedding of HLA-DRB1 alleles<a class="anchor" aria-label="anchor" href="#umap-embedding-of-hla-drb1-alleles"></a>
</h2>
<p>There are many possibilities for further analysis of the one-hot
encoding matrix.</p>
<p>For example, here is a UMAP embedding of HLA-DRB1 alleles encoded as
a one-hot amino acid matrix with 1725 columns, one for each amino acid
at each position. The color indicates the 2-digit allele name.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">uamp</span><span class="op">(</span><span class="va">a</span><span class="op">$</span><span class="va">onehot</span>, n_epochs <span class="op">=</span> <span class="fl">200</span>, min_dist <span class="op">=</span> <span class="fl">1</span>, spread <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="examples_files/figure-html/umap-2digit-1.png" width="2700"></p>
<p>We can highlight which alleles have aspartic acid (Asp or D) at
position 57:</p>
<p><img src="examples_files/figure-html/umap-D57-1.png" width="2700"></p>
<p>Or we can use color to represent the amino acid residue at position
57:</p>
<p><img src="examples_files/figure-html/umap-57-1.png" width="2700"></p>
</div>
<div class="section level2">
<h2 id="get-hla-allele-frequencies-from-allele-frequency-net-database-afnd">Get HLA allele frequencies from Allele Frequency Net Database
(AFND)<a class="anchor" aria-label="anchor" href="#get-hla-allele-frequencies-from-allele-frequency-net-database-afnd"></a>
</h2>
<p>The hlabud R package includes a table of HLA allele frequencies from
the <a href="http://www.allelefrequencies.net/" class="external-link">Allele Frequency Net
Database (AFND)</a>.</p>
<p>If you use this data, please cite the latest manuscript about Allele
Frequency Net Database:</p>
<ul>
<li>Gonzalez-Galarza FF, McCabe A, Santos EJMD, Jones J, Takeshita L,
Ortega-Rivera ND, et al. <a href="https://pubmed.ncbi.nlm.nih.gov/31722398" class="external-link">Allele frequency net
database (AFND) 2020 update: gold-standard data classification, open
access genotype data and new query tools.</a> Nucleic Acids Res.
2020;48: D783–D788. <a href="doi:10.1093/nar/gkz1029" class="uri">doi:10.1093/nar/gkz1029</a>
</li>
</ul>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">af</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hla_frequencies.html">hla_frequencies</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">af</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 123,502 × 7</span></span></span>
<span><span class="co">#&gt;    group gene  allele  population            indivs_over_n alleles_over_2n     n</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> hla   A     A*01:01 Argentina Rosario To…          15.1           0.076    86</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> hla   A     A*01:01 Armenia combined Reg…          <span style="color: #BB0000;">NA</span>             0.125   100</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> hla   A     A*01:01 Australia Cape York …          <span style="color: #BB0000;">NA</span>             0.053   103</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> hla   A     A*01:01 Australia Groote Eyl…          <span style="color: #BB0000;">NA</span>             0.027    75</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> hla   A     A*01:01 Australia New South …          <span style="color: #BB0000;">NA</span>             0.187   134</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> hla   A     A*01:01 Australia Yuendumu A…          <span style="color: #BB0000;">NA</span>             0.008   191</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> hla   A     A*01:01 Austria                        27             0.146   200</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> hla   A     A*01:01 Azores Central Islan…          <span style="color: #BB0000;">NA</span>             0.08     59</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> hla   A     A*01:01 Azores Oriental Isla…          <span style="color: #BB0000;">NA</span>             0.115    43</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> hla   A     A*01:01 Azores Terceira Isla…          <span style="color: #BB0000;">NA</span>             0.109   130</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 123,492 more rows</span></span></span></code></pre></div>
<p>We can use this data to plot the frequency of a specific allele
(e.g. DQB1*02:01) in populations with more than 1000 sampled
individuals:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">my_allele</span> <span class="op">&lt;-</span> <span class="st">"DQB1*02:01"</span></span>
<span><span class="va">my_af</span> <span class="op">&lt;-</span> <span class="va">af</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">allele</span> <span class="op">==</span> <span class="va">my_allele</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">n</span> <span class="op">&gt;</span> <span class="fl">1000</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">arrange</span><span class="op">(</span><span class="op">-</span><span class="va">alleles_over_2n</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="va">my_af</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">alleles_over_2n</span>, y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/reorder.factor.html" class="external-link">reorder</a></span><span class="op">(</span><span class="va">population</span>, <span class="va">alleles_over_2n</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">scale_y_discrete</span><span class="op">(</span>position <span class="op">=</span> <span class="st">"right"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_colh</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span></span>
<span>    x <span class="op">=</span> <span class="st">"Allele Frequency (Alleles / 2N)"</span>,</span>
<span>    y <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>    title <span class="op">=</span>  <span class="fu">glue</span><span class="op">(</span><span class="st">"Frequency of {my_allele} across {length(unique(my_af$population))} populations"</span><span class="op">)</span>,</span>
<span>    caption <span class="op">=</span> <span class="st">"Data from AFND http://allelefrequencies.net"</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="examples_files/figure-html/afnd_dqb1_02_01-1.png" width="3000"></p>
<p>See <a href="https://github.com/slowkow/allelefrequencies" class="external-link">github.com/slowkow/allelefrequencies</a>
for more examples of how we might use this data.</p>
</div>
<div class="section level2">
<h2 id="compute-hla-divergence-with-the-grantham-distance-matrix">Compute HLA divergence with the Grantham distance matrix<a class="anchor" aria-label="anchor" href="#compute-hla-divergence-with-the-grantham-distance-matrix"></a>
</h2>
<p>Humans are diploid, so each of us has two copies of each HLA gene. An
individual with two highly dissimilar alleles can bind a greater number
of different peptides than a homozygous individual (<a href="https://doi.org/10.1007/BF02918202" class="external-link uri">https://doi.org/10.1007/BF02918202</a>):</p>
<blockquote>
<p>Each MHC class II allele has the capacity to bind and present a
specific set of peptides from processed antigens. The inability of a
specific class II allele to bind and present a fragment derived from a
processed antigen results in the loss of immune responsiveness for that
antigen in individuals homozygous for that class II allele.</p>
</blockquote>
<p>The amino acid distance matrix by Granthan 1974 (<a href="https://doi.org/10.1126/science.185.4154.862" class="external-link uri">https://doi.org/10.1126/science.185.4154.862</a>) encodes
information about the composition, polarity, and molecular volume of
each amino acid.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">grantham</span></span>
<span><span class="co">#&gt;    amino    c    p     v</span></span>
<span><span class="co">#&gt; 1    Ser 1.42  9.2  32.0</span></span>
<span><span class="co">#&gt; 2    Arg 0.65 10.5 124.0</span></span>
<span><span class="co">#&gt; 3    Leu 0.00  4.9 111.0</span></span>
<span><span class="co">#&gt; 4    Pro 0.39  8.0  32.5</span></span>
<span><span class="co">#&gt; 5    Thr 0.71  8.6  61.0</span></span>
<span><span class="co">#&gt; 6    Ala 0.00  8.1  31.0</span></span>
<span><span class="co">#&gt; 7    Val 0.00  5.9  84.0</span></span>
<span><span class="co">#&gt; 8    Gly 0.74  9.0   3.0</span></span>
<span><span class="co">#&gt; 9    Ile 0.00  5.2 111.0</span></span>
<span><span class="co">#&gt; 10   Phe 0.00  5.2 132.0</span></span>
<span><span class="co">#&gt; 11   Tyr 0.20  6.2 136.0</span></span>
<span><span class="co">#&gt; 12   Cys 2.75  5.5  55.0</span></span>
<span><span class="co">#&gt; 13   His 0.58 10.4  96.0</span></span>
<span><span class="co">#&gt; 14   Gln 0.89 10.5  85.0</span></span>
<span><span class="co">#&gt; 15   Asn 1.33 11.6  56.0</span></span>
<span><span class="co">#&gt; 16   Lys 0.33 11.3 119.0</span></span>
<span><span class="co">#&gt; 17   Asp 1.38 13.0  54.0</span></span>
<span><span class="co">#&gt; 18   Glu 0.92 12.3  83.0</span></span>
<span><span class="co">#&gt; 19   Met 0.00  5.7 105.0</span></span>
<span><span class="co">#&gt; 20   Trp 0.13  5.4 170.0</span></span></code></pre></div>
<p>We can use that matrix to compute an HLA divergence metric for a set
of individuals like this:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">my_genos</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A*23:01:12,A*24:550"</span>, <span class="st">"A*25:12N,A*11:27"</span>, <span class="st">"A*24:381,A*33:85"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/hla_divergence.html">hla_divergence</a></span><span class="op">(</span><span class="va">my_genos</span><span class="op">)</span></span>
<span><span class="co">#&gt; A*23:01:12,A*24:550    A*25:12N,A*11:27    A*24:381,A*33:85 </span></span>
<span><span class="co">#&gt;            0.511811            3.464567            5.094488</span></span></code></pre></div>
<p>The divergence for a homozygote is equal to zero, by definition:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/hla_divergence.html">hla_divergence</a></span><span class="op">(</span><span class="st">"A*01:01,A*01:01"</span><span class="op">)</span></span>
<span><span class="co">#&gt; A*01:01,A*01:01 </span></span>
<span><span class="co">#&gt;               0</span></span></code></pre></div>
<p>hlabud includes R code for the divergence calculations that was
translated from the original Perl code by Pierini &amp; Lenz 2018 (<a href="https://doi.org/10.1093/molbev/msy116" class="external-link uri">https://doi.org/10.1093/molbev/msy116</a>).</p>
<p>The amino acid distance matrix is easily accessible, and we provide
two built-in options “grantham” and “uniform”:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/amino_distance_matrix.html">amino_distance_matrix</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"grantham"</span><span class="op">)</span></span>
<span><span class="co">#&gt;     A   R   N   D   C   Q   E   G   H   I   L   K   M   F   P   S   T   W   Y</span></span>
<span><span class="co">#&gt; A   0 112 111 126 195  91 107  60  86  94  96 106  84 113  27  99  58 148 112</span></span>
<span><span class="co">#&gt; R 112   0  86  96 180  43  54 125  29  97 102  26  91  97 103 110  71 101  77</span></span>
<span><span class="co">#&gt; N 111  86   0  23 139  46  42  80  68 149 153  94 142 158  91  46  65 174 143</span></span>
<span><span class="co">#&gt; D 126  96  23   0 154  61  45  94  81 168 172 101 160 177 108  65  85 181 160</span></span>
<span><span class="co">#&gt; C 195 180 139 154   0 154 170 159 174 198 198 202 196 205 169 112 149 215 194</span></span>
<span><span class="co">#&gt; Q  91  43  46  61 154   0  29  87  24 109 113  53 101 116  76  68  42 130  99</span></span>
<span><span class="co">#&gt; E 107  54  42  45 170  29   0  98  40 134 138  56 126 140  93  80  65 152 122</span></span>
<span><span class="co">#&gt; G  60 125  80  94 159  87  98   0  98 135 138 127 127 153  42  56  59 184 147</span></span>
<span><span class="co">#&gt; H  86  29  68  81 174  24  40  98   0  94  99  32  87 100  77  89  47 115  83</span></span>
<span><span class="co">#&gt; I  94  97 149 168 198 109 134 135  94   0   5 102  10  21  95 142  89  61  33</span></span>
<span><span class="co">#&gt; L  96 102 153 172 198 113 138 138  99   5   0 107  15  22  98 145  92  61  36</span></span>
<span><span class="co">#&gt; K 106  26  94 101 202  53  56 127  32 102 107   0  95 102 103 121  78 110  85</span></span>
<span><span class="co">#&gt; M  84  91 142 160 196 101 126 127  87  10  15  95   0  28  87 135  81  67  36</span></span>
<span><span class="co">#&gt; F 113  97 158 177 205 116 140 153 100  21  22 102  28   0 114 155 103  40  22</span></span>
<span><span class="co">#&gt; P  27 103  91 108 169  76  93  42  77  95  98 103  87 114   0  74  38 147 110</span></span>
<span><span class="co">#&gt; S  99 110  46  65 112  68  80  56  89 142 145 121 135 155  74   0  58 177 144</span></span>
<span><span class="co">#&gt; T  58  71  65  85 149  42  65  59  47  89  92  78  81 103  38  58   0 128  92</span></span>
<span><span class="co">#&gt; W 148 101 174 181 215 130 152 184 115  61  61 110  67  40 147 177 128   0  37</span></span>
<span><span class="co">#&gt; Y 112  77 143 160 194  99 122 147  83  33  36  85  36  22 110 144  92  37   0</span></span>
<span><span class="co">#&gt; V  64  96 133 152 192  96 121 109  84  29  32  97  21  50  68 124  69  88  55</span></span>
<span><span class="co">#&gt;     V</span></span>
<span><span class="co">#&gt; A  64</span></span>
<span><span class="co">#&gt; R  96</span></span>
<span><span class="co">#&gt; N 133</span></span>
<span><span class="co">#&gt; D 152</span></span>
<span><span class="co">#&gt; C 192</span></span>
<span><span class="co">#&gt; Q  96</span></span>
<span><span class="co">#&gt; E 121</span></span>
<span><span class="co">#&gt; G 109</span></span>
<span><span class="co">#&gt; H  84</span></span>
<span><span class="co">#&gt; I  29</span></span>
<span><span class="co">#&gt; L  32</span></span>
<span><span class="co">#&gt; K  97</span></span>
<span><span class="co">#&gt; M  21</span></span>
<span><span class="co">#&gt; F  50</span></span>
<span><span class="co">#&gt; P  68</span></span>
<span><span class="co">#&gt; S 124</span></span>
<span><span class="co">#&gt; T  69</span></span>
<span><span class="co">#&gt; W  88</span></span>
<span><span class="co">#&gt; Y  55</span></span>
<span><span class="co">#&gt; V   0</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="download-and-unpack-all-data-from-the-latest-imgthla-release">Download and unpack all data from the latest IMGTHLA release<a class="anchor" aria-label="anchor" href="#download-and-unpack-all-data-from-the-latest-imgthla-release"></a>
</h2>
<p>If you only want to use <code><a href="../reference/hla_alignments.html">hla_alignments()</a></code>, then you don’t
need <code><a href="../reference/install_hla.html">install_hla()</a></code> because data files are downloaded
automatically as needed and cached for future use.</p>
<p>But some users might need access to additional files that are only
present in the full data release.</p>
<p>Run <code><a href="../reference/install_hla.html">install_hla()</a></code> to download and unpack the <a href="https://github.com/ANHIG/IMGTHLA/releases" class="external-link">latest IMGTHLA
release</a>. The destination folder for the downloaded data files is
<code>getOption("hlabud_dir")</code> (automatically tailored to your
operating system thanks to the <a href="https://github.com/r-lib/rappdirs" class="external-link">rappdirs</a> package).</p>
<p>Here are a few examples of how to download releases or get a list of
release names.</p>
<p>Download the latest release (default) or a specific release:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Download all of the data (120MB) for the latest IMGTHLA release</span></span>
<span><span class="fu"><a href="../reference/install_hla.html">install_hla</a></span><span class="op">(</span>release <span class="op">=</span> <span class="st">"latest"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Download a specific release</span></span>
<span><span class="fu"><a href="../reference/install_hla.html">install_hla</a></span><span class="op">(</span>release <span class="op">=</span> <span class="st">"3.51.0"</span><span class="op">)</span></span></code></pre></div>
<p>Optionally, get or set the directory hlabud uses to store the
data:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"hlabud_dir"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "/home/username/.local/share/hlabud"</span></span>
<span></span>
<span><span class="co"># Manually override the directory for hlabud to use</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>hlabud_dir <span class="op">=</span> <span class="st">"/path/to/my/dir"</span><span class="op">)</span></span></code></pre></div>
<p>After installing a few releases, the hlabud folder might look like
this:</p>
<pre><code>❯ ls -lah "/home/user/.local/share/hlabud"
total 207M
drwxrwxr-x  3 user user      32 Apr  5 01:19 3.30.0
drwxrwxr-x 11 user user    4.0K Apr  7 19:31 3.40.0
drwxrwxr-x 12 user user    4.0K Apr  5 00:27 3.51.0
-rw-rw-r--  1 user user     15K Apr  7 19:23 tags.json
-rw-rw-r--  1 user user     79M Apr  7 19:28 v3.40.0-alpha.tar.gz
-rw-rw-r--  1 user user    129M Apr  4 20:07 v3.51.0-alpha.tar.gz</code></pre>
</div>
<div class="section level2">
<h2 id="count-the-number-of-alleles-in-each-imgthla-release">Count the number of alleles in each IMGTHLA release<a class="anchor" aria-label="anchor" href="#count-the-number-of-alleles-in-each-imgthla-release"></a>
</h2>
<p>We can get a list of the release names:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">releases</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hla_releases.html">hla_releases</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">releases</span></span>
<span><span class="co">#&gt;  [1] "3.59.0"   "3.58.0"   "3.57.0"   "3.56.0"   "3.55.0"   "3.54.0"  </span></span>
<span><span class="co">#&gt;  [7] "3.53.0"   "3.52.0"   "3.51.0"   "3.50.0"   "3.49.0"   "3.48.0"  </span></span>
<span><span class="co">#&gt; [13] "3.47.0"   "3.46.0"   "3.45.1"   "3.45.01"  "3.45.0.1" "3.45.0"  </span></span>
<span><span class="co">#&gt; [19] "3.44.1"   "3.44.0"   "3.43.0"   "3.42.0"   "3.41.2"   "3.41.0"  </span></span>
<span><span class="co">#&gt; [25] "3.40.0"   "3.39.0"   "3.38.0"   "3.37.0"   "3.36.0"   "3.35.0"</span></span></code></pre></div>
<p>Then we can get the allele names for each release:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">my_alleles</span> <span class="op">&lt;-</span> <span class="fu">rbindlist</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">releases</span>, <span class="kw">function</span><span class="op">(</span><span class="va">release</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">retval</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hla_alleles.html">hla_alleles</a></span><span class="op">(</span>release <span class="op">=</span> <span class="va">release</span><span class="op">)</span></span>
<span>  <span class="va">retval</span><span class="op">$</span><span class="va">release</span> <span class="op">&lt;-</span> <span class="va">release</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">retval</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span>, fill <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in hla_alleles(release = release): unrecognized release name</span></span>
<span><span class="co">#&gt; 'Allelelist.3451.txt'</span></span>
<span><span class="co">#&gt; Warning in hla_alleles(release = release): unrecognized release name</span></span>
<span><span class="co">#&gt; 'Allelelist.34501.txt'</span></span>
<span><span class="co">#&gt; Warning in hla_alleles(release = release): unrecognized release name</span></span>
<span><span class="co">#&gt; 'Allelelist.34501.txt'</span></span>
<span><span class="co">#&gt; Warning in hla_alleles(release = release): unrecognized release name</span></span>
<span><span class="co">#&gt; 'Allelelist.3441.txt'</span></span>
<span><span class="co">#&gt; Warning in hla_alleles(release = release): unrecognized release name</span></span>
<span><span class="co">#&gt; 'Allelelist.3412.txt'</span></span></code></pre></div>
<p>Next, count how many alleles we have in each release:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">d</span> <span class="op">&lt;-</span> <span class="va">my_alleles</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu">count</span><span class="op">(</span><span class="va">release</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">n</span> <span class="op">&gt;</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">d</span></span>
<span><span class="co">#&gt;     release     n</span></span>
<span><span class="co">#&gt;      &lt;char&gt; &lt;int&gt;</span></span>
<span><span class="co">#&gt;  1:  3.35.0 21683</span></span>
<span><span class="co">#&gt;  2:  3.36.0 22548</span></span>
<span><span class="co">#&gt;  3:  3.37.0 24093</span></span>
<span><span class="co">#&gt;  4:  3.38.0 25958</span></span>
<span><span class="co">#&gt;  5:  3.39.0 26512</span></span>
<span><span class="co">#&gt;  6:  3.40.0 27273</span></span>
<span><span class="co">#&gt;  7:  3.41.0 27980</span></span>
<span><span class="co">#&gt;  8:  3.42.0 28786</span></span>
<span><span class="co">#&gt;  9:  3.43.0 29417</span></span>
<span><span class="co">#&gt; 10:  3.44.0 30523</span></span>
<span><span class="co">#&gt; 11:  3.45.0 31552</span></span>
<span><span class="co">#&gt; 12:  3.46.0 32330</span></span>
<span><span class="co">#&gt; 13:  3.47.0 33552</span></span>
<span><span class="co">#&gt; 14:  3.48.0 34145</span></span>
<span><span class="co">#&gt; 15:  3.49.0 35077</span></span>
<span><span class="co">#&gt; 16:  3.50.0 36016</span></span>
<span><span class="co">#&gt; 17:  3.51.0 36625</span></span>
<span><span class="co">#&gt; 18:  3.52.0 37068</span></span>
<span><span class="co">#&gt; 19:  3.53.0 37619</span></span>
<span><span class="co">#&gt; 20:  3.54.0 38416</span></span>
<span><span class="co">#&gt; 21:  3.55.0 38909</span></span>
<span><span class="co">#&gt; 22:  3.56.0 39886</span></span>
<span><span class="co">#&gt; 23:  3.57.0 40623</span></span>
<span><span class="co">#&gt; 24:  3.58.0 41433</span></span>
<span><span class="co">#&gt; 25:  3.59.0 42022</span></span>
<span><span class="co">#&gt;     release     n</span></span></code></pre></div>
<p>And plot the number of alleles as a line plot:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span><span class="va">d</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">release</span>, y <span class="op">=</span> <span class="va">n</span>, group <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_line</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_text</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>label <span class="op">=</span> <span class="va">release</span><span class="op">)</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span>x <span class="op">=</span> <span class="cn">NULL</span>, y <span class="op">=</span> <span class="st">"Number of alleles"</span>,</span>
<span>  title <span class="op">=</span> <span class="st">"Each release has more HLA alleles"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span></span>
<span>    axis.text.x <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>,</span>
<span>    axis.ticks.x <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="examples_files/figure-html/alleles_per_release-1.png" width="700"></p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">d2</span> <span class="op">&lt;-</span> <span class="va">my_alleles</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu">mutate</span><span class="op">(</span>gene <span class="op">=</span> <span class="fu">str_split_fixed</span><span class="op">(</span><span class="va">Allele</span>, <span class="st">"\\*"</span>, <span class="fl">2</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu">count</span><span class="op">(</span><span class="va">release</span>, <span class="va">gene</span><span class="op">)</span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">release</span>, y <span class="op">=</span> <span class="va">n</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_line</span><span class="op">(</span></span>
<span>    data <span class="op">=</span> <span class="va">d2</span>,</span>
<span>    <span class="fu">aes</span><span class="op">(</span>group <span class="op">=</span> <span class="va">gene</span>, color <span class="op">=</span> <span class="va">gene</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">scale_color_discrete</span><span class="op">(</span>guide <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_text</span><span class="op">(</span></span>
<span>    data <span class="op">=</span> <span class="va">d2</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">release</span> <span class="op">==</span> <span class="st">"3.52.0"</span><span class="op">)</span>,</span>
<span>    mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>label <span class="op">=</span> <span class="va">gene</span><span class="op">)</span>,</span>
<span>    hjust <span class="op">=</span> <span class="fl">0</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span>x <span class="op">=</span> <span class="cn">NULL</span>, y <span class="op">=</span> <span class="st">"Number of alleles"</span>,</span>
<span>  title <span class="op">=</span> <span class="st">"Number of alleles per release and gene"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">scale_x_discrete</span><span class="op">(</span>expand <span class="op">=</span> <span class="fu">expansion</span><span class="op">(</span>mult <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.01</span>, <span class="fl">0.1</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">scale_y_log10</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span></span>
<span>    panel.grid.major.y <span class="op">=</span> <span class="fu">element_line</span><span class="op">(</span><span class="op">)</span>, </span>
<span>    axis.text.x <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>,</span>
<span>    axis.ticks.x <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="examples_files/figure-html/alleles_per_release_and_gene-1.png" width="700"></p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Kamil Slowikowski.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
