<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="hlabud">
<title>Numbering amino acid positions • hlabud</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Numbering amino acid positions">
<meta property="og:description" content="hlabud">
<meta name="twitter:card" content="summary_large_image">
<meta name="twitter:creator" content="@slowkow">
<meta name="twitter:site" content="@slowkow">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">hlabud</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/examples.html">hlabud usage examples</a>
    <a class="dropdown-item" href="../articles/numbering.html">Numbering amino acid positions</a>
    <a class="dropdown-item" href="../articles/visualize-hla-structure.html">Visualize HLA protein structures</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/slowkow/hlabud">
    <span class="fa fa-github"></span>
     
    GitHub
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Numbering amino acid positions</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/slowkow/hlabud/blob/HEAD/vignettes/articles/numbering.Rmd" class="external-link"><code>vignettes/articles/numbering.Rmd</code></a></small>
      <div class="d-none name"><code>numbering.Rmd</code></div>
    </div>

    
    
<div class="section level3">
<h3 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h3>
<p>Kamil Slowikowski</p>
<p>2024-05-15</p>
<p>The <a href="https://github.com/ANHIG/IMGTHLA" class="external-link">IMGTHLA</a> provides a
Github repo with alignments of amino acid sequences and nucleotide
sequences for thousands of alleles of the HLA genes. The IMGTHLA
alignments define the official numbering scheme, and they provide some
explanations for the conventions on their <a href="https://www.ebi.ac.uk/ipd/mhc/alignment/help/" class="external-link">help page</a>.</p>
<p>The <a href="https://github.com/slowkow/hlabud" class="external-link">hlabud</a> R package
provides easy access to the alignment data, and hlabud follows the
official numbering scheme. The examples below should help beginners to
visualize and understand how the conventions work.</p>
</div>
<div class="section level2">
<h2 id="alignment-files-on-the-imgthla-github-page">Alignment files on the IMGTHLA Github page<a class="anchor" aria-label="anchor" href="#alignment-files-on-the-imgthla-github-page"></a>
</h2>
<p>The IMGTHLA Github page provides <a href="https://github.com/ANHIG/IMGTHLA/tree/v3.56.0-alpha/alignments" class="external-link">a
folder</a> with alignment files.</p>
<p><img src="https://github.com/ANHIG/IMGTHLA/assets/209714/b0eb1c09-6a49-46f3-b8d9-21db46a82059"></p>
<p>For the examples in this vignette, we will use the <em>HLA-DRB1</em>
gene.</p>
<p>For DRB1, we can find three separate files:</p>
<ul>
<li><a href="https://raw.githubusercontent.com/ANHIG/IMGTHLA/v3.56.0-alpha/alignments/DRB1_gen.txt" class="external-link uri">https://raw.githubusercontent.com/ANHIG/IMGTHLA/v3.56.0-alpha/alignments/DRB1_gen.txt</a></li>
<li><a href="https://raw.githubusercontent.com/ANHIG/IMGTHLA/v3.56.0-alpha/alignments/DRB1_nuc.txt" class="external-link uri">https://raw.githubusercontent.com/ANHIG/IMGTHLA/v3.56.0-alpha/alignments/DRB1_nuc.txt</a></li>
<li><a href="https://raw.githubusercontent.com/ANHIG/IMGTHLA/v3.56.0-alpha/alignments/DRB1_prot.txt" class="external-link uri">https://raw.githubusercontent.com/ANHIG/IMGTHLA/v3.56.0-alpha/alignments/DRB1_prot.txt</a></li>
</ul>
<p>The files contain different information:</p>
<ul>
<li>
<code>gen</code> contains genomic DNA sequences.</li>
<li>
<code>nuc</code> contains nucleotide coding sequences (CDS).</li>
<li>
<code>prot</code> contains protein sequences (amino acids).</li>
</ul>
<p>Let’s consider the <code>DRB1_prot.txt</code> file. What does the
file look like?</p>
<p><img src="https://github.com/ANHIG/IMGTHLA/assets/209714/5ee93946-a31d-4072-883d-daa92bf448a7"></p>
<p>It is a plain text file with a header and sequence alignments. In the
alignment, each line represents one sequence (or allele), and each line
has 100 residues. The first 100 residues for all alleles are shown in
the first block. Then, the next block has the next 100 residues for all
of the alleles, and so on.</p>
<div class="section level3">
<h3 id="numbering-conventions">Numbering conventions<a class="anchor" aria-label="anchor" href="#numbering-conventions"></a>
</h3>
<p>Here are the conventions used for alignments (copied from EBI):</p>
<blockquote>
<ul>
<li>The entry for each allele is displayed in respect to the reference
sequences.</li>
<li>Where identity to the reference sequence is present the base will be
displayed as a hyphen (<code>-</code>).</li>
<li>Non-identity to the reference sequence is shown by displaying the
appropriate base at that position.</li>
<li>Where an insertion or deletion has occurred this will be represented
by a period (<code>.</code>).</li>
<li>If the sequence is unknown at any point in the alignment, this will
be represented by an asterisk (<code>*</code>).</li>
<li>In protein alignments for null alleles, the ‘Stop’ codons will be
represented by a hash (<code>X</code>).</li>
<li>In protein alignments, sequence following the termination codon,
will not be marked and will appear blank.</li>
<li>These conventions are used for both nucleotide and protein
alignments.</li>
</ul>
</blockquote>
<p>That’s a lot of information! Let’s try to work through an example to
illustrate how this works.</p>
<p>The first sequence in the alignment is the <strong>reference
sequence</strong>. The position numbering is relative to the reference
sequence. That means deletions (<code>.</code>) in the reference
sequence are not numbered.</p>
<p>Notice below that the numbering starts with negative numbers. The <a href="https://www.ebi.ac.uk/ipd/mhc/alignment/help/" class="external-link">help page</a>
clarifies:</p>
<blockquote>
<p>Protein Sequence Numbering</p>
<ul>
<li>For amino acid-based systems, the start codon of the mature protein
is labeled codon 1.</li>
<li>The codon 5’ to this is numbered -1.</li>
<li>All numbering is based on the reference sequence.</li>
</ul>
</blockquote>
<p>There is no amino acid with the number 0.</p>
</div>
<div class="section level3">
<h3 id="numbering-indels">Numbering indels<a class="anchor" aria-label="anchor" href="#numbering-indels"></a>
</h3>
<p>The alignment below shows that 100 residues are displayed in chunks
of 10:</p>
<p><img src="https://github.com/ANHIG/IMGTHLA/assets/209714/0e9d425b-a6fe-4a3a-9aa2-285698087f26"></p>
<p>The numbering convention says that indels in the reference sequence
are not numbered.</p>
<p>To clarify this point, I manually added additional numbers (11, 21,
30, 39, 49, 59) to the alignment below:</p>
<p><img src="https://github.com/ANHIG/IMGTHLA/assets/209714/64666703-0ab2-4280-a776-a8c39195d209"></p>
<p>Notice that when we move from the first chunk <code>GDTRPRFLWQ</code>
to the next chunk <code>LKFECHFFNG</code> we simply add 10 to 1 to get
11 for the number of the <code>L</code> amino acid. Then, as we move on
to <code>TERVR.LLER</code>, we add 10 to 11 to get 21 for the
<code>T</code> amino acid.</p>
<p>However, as we move on to <code>CIYNQEE.SV</code> the rule of “add
10” does not work. Instead of labeling <code>C</code> as position 31, we
label it position 30. Why?</p>
<p>The reason why <code>C</code> is 31, and not 30, is because there is
an indel (or gap) in the reference sequence at position 25_26 (notice
the <code>.</code> in <code>R.L</code>). The convention is that
deletions in the reference sequence are not numbered.</p>
<p>Let’s take a closer look at this data with hlabud.</p>
<p>Here are the first few amino acid positions for the first 4
sequences:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/slowkow/hlabud" class="external-link">hlabud</a></span><span class="op">)</span></span>
<span><span class="va">a</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hla_alignments.html">hla_alignments</a></span><span class="op">(</span><span class="st">"DRB1"</span>, release <span class="op">=</span> <span class="st">"3.56.0"</span><span class="op">)</span></span>
<span><span class="va">seqs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/substr.html" class="external-link">substr</a></span><span class="op">(</span><span class="va">a</span><span class="op">$</span><span class="va">sequences</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span>, <span class="fl">30</span>, <span class="fl">89</span><span class="op">)</span></span>
<span><span class="fu">str_replace_all</span><span class="op">(</span><span class="va">seqs</span>, <span class="st">"(\\S{10})"</span>, <span class="st">"\\1 "</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "GDTRPRFLWQ LKFECHFFNG TERVR.LLER CIYNQEE.SV RFDSDVGEYR AVTELGRPDA "</span></span>
<span><span class="co">#&gt; [2] "---------- ---------- -----.---- -------.-- ---------- ---------- "</span></span>
<span><span class="co">#&gt; [3] "---------- ---------- -----.---- -------.-- ---------- ---------- "</span></span>
<span><span class="co">#&gt; [4] "---------- ---------- -----.---- -------.-- ---------- ---------- "</span></span></code></pre></div>
<p>This is how hlabud numbers the positions that we are focusing on in
this example:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">a</span><span class="op">$</span><span class="va">alleles</span><span class="op">)</span><span class="op">[</span><span class="fl">50</span><span class="op">:</span><span class="fl">70</span><span class="op">]</span></span>
<span><span class="co">#&gt;  [1] "21"    "22"    "23"    "24"    "25"    "25_26" "26"    "27"    "28"   </span></span>
<span><span class="co">#&gt; [10] "29"    "30"    "31"    "32"    "33"    "34"    "35"    "36"    "36_37"</span></span>
<span><span class="co">#&gt; [19] "37"    "38"    "39"</span></span></code></pre></div>
<p>If hlabud is using the correct numbering, then we should see: -
<code>T</code> at position 21 - <code>C</code> at position 30</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">a</span><span class="op">$</span><span class="va">alleles</span><span class="op">[</span><span class="fl">1</span>,<span class="st">"21"</span><span class="op">]</span></span>
<span><span class="co">#&gt; [1] "T"</span></span>
<span><span class="va">a</span><span class="op">$</span><span class="va">alleles</span><span class="op">[</span><span class="fl">1</span>,<span class="st">"30"</span><span class="op">]</span></span>
<span><span class="co">#&gt; [1] "C"</span></span></code></pre></div>
<p>What do we see at positions 25, 26, and 25_26?</p>
<p>Here is the alignment file:</p>
<p><img src="https://github.com/ANHIG/IMGTHLA/assets/209714/7a779b7f-2d2c-4817-95b7-986c2c70171c"></p>
<p>And here is the result from hlabud:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">a</span><span class="op">$</span><span class="va">alleles</span><span class="op">[</span><span class="fl">1</span>,<span class="st">"25"</span><span class="op">]</span></span>
<span><span class="co">#&gt; [1] "R"</span></span>
<span><span class="va">a</span><span class="op">$</span><span class="va">alleles</span><span class="op">[</span><span class="fl">1</span>,<span class="st">"26"</span><span class="op">]</span></span>
<span><span class="co">#&gt; [1] "L"</span></span>
<span><span class="va">a</span><span class="op">$</span><span class="va">alleles</span><span class="op">[</span><span class="fl">1</span>,<span class="st">"25_26"</span><span class="op">]</span></span>
<span><span class="co">#&gt; [1] "."</span></span></code></pre></div>
<p>So, we can see that the deletion between positions 25 and 26 is not
numbered like the other residues. Instead, it gets a special label
(25_26) that consists of the positions flanking the indel (25 and
26).</p>
<p>What alleles do we observe at position 25_26?</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">a</span><span class="op">$</span><span class="va">alleles</span><span class="op">[</span>,<span class="st">"25_26"</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;    .    *    W </span></span>
<span><span class="co">#&gt; 3658   12    1</span></span></code></pre></div>
<p>There are three possibilities at position 25_26:</p>
<ul>
<li>
<code>.</code> indicates a deletion of 1 amino acid (or the absence
of an amino at this position)</li>
<li>
<code>*</code> indicates that the sequence is unknown at this
position</li>
<li>
<code>W</code> indicates tryptophan at this position</li>
</ul>
<p>I hope this example helps to explain the numbering of indels.</p>
<p>If you notice any discrepancy between hlabud and IMGT, please <a href="https://github.com/slowkow/hlabud/issues" class="external-link">report it</a>.</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Kamil Slowikowski.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
